<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Module 3 - Lab practical | Advanced Microbiome Analysis 2025</title>
  <meta name="description" content="CBW’s Bookdown Template for Workshops" />
  <meta name="generator" content="bookdown 0.43 and GitBook 2.6.7" />

  <meta property="og:title" content="Module 3 - Lab practical | Advanced Microbiome Analysis 2025" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="https://cbw-dev.github.io/bookdown-template/img/bioinformatics_logo.png" />
  <meta property="og:description" content="CBW’s Bookdown Template for Workshops" />
  <meta name="github-repo" content="cbw-dev/AMB_2025" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Module 3 - Lab practical | Advanced Microbiome Analysis 2025" />
  
  <meta name="twitter:description" content="CBW’s Bookdown Template for Workshops" />
  <meta name="twitter:image" content="https://cbw-dev.github.io/bookdown-template/img/bioinformatics_logo.png" />




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="img/favicon.ico" type="image/x-icon" />
<link rel="prev" href="module-3-assigning-functions.html"/>
<link rel="next" href="module-4-visualization-and-finding-functional-significance.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<a href="https://bioinformaticsdotca.github.io/">
 <img src="img/bioinformatics_logo.png" width="90%" alt="bioinformatics.ca logo" style="display:block; margin-left:auto; margin-right:auto; margin-top:15px;">
 </a>

 <li><a href="./">Advanced Microbiome Analysis 2025</a></li>

<li class="divider"></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="chapter" data-level="" data-path="course-schedule.html"><a href="course-schedule.html"><i class="fa fa-check"></i>Course Schedule</a>
<ul>
<li class="chapter" data-level="" data-path="course-schedule.html"><a href="course-schedule.html#meet-your-faculty"><i class="fa fa-check"></i>Meet Your Faculty</a></li>
</ul></li>
<li class="part"><span><b>II Modules</b></span></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html"><i class="fa fa-check"></i>Module 1</a></li>
<li class="chapter" data-level="" data-path="module-3-assigning-functions.html"><a href="module-3-assigning-functions.html"><i class="fa fa-check"></i>Module 3: Assigning Functions</a></li>
<li class="chapter" data-level="" data-path="module-3---lab-practical.html"><a href="module-3---lab-practical.html"><i class="fa fa-check"></i>Module 3 - Lab practical</a>
<ul>
<li class="chapter" data-level="" data-path="module-3---lab-practical.html"><a href="module-3---lab-practical.html#setting-up-the-workspace"><i class="fa fa-check"></i>Setting up the workspace</a>
<ul>
<li class="chapter" data-level="" data-path="module-3---lab-practical.html"><a href="module-3---lab-practical.html#ssh-port-forwarding"><i class="fa fa-check"></i>SSH port forwarding</a></li>
<li class="chapter" data-level="" data-path="module-3---lab-practical.html"><a href="module-3---lab-practical.html#workspace"><i class="fa fa-check"></i>Workspace</a></li>
<li class="chapter" data-level="" data-path="module-3---lab-practical.html"><a href="module-3---lab-practical.html#start-container"><i class="fa fa-check"></i>Start container</a></li>
<li class="chapter" data-level="" data-path="module-3---lab-practical.html"><a href="module-3---lab-practical.html#jupyter"><i class="fa fa-check"></i>Jupyter</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-3---lab-practical.html"><a href="module-3---lab-practical.html#overview"><i class="fa fa-check"></i>Overview</a></li>
<li class="chapter" data-level="" data-path="module-3---lab-practical.html"><a href="module-3---lab-practical.html#prodigal"><i class="fa fa-check"></i>Prodigal</a></li>
<li class="chapter" data-level="" data-path="module-3---lab-practical.html"><a href="module-3---lab-practical.html#sequence-similarity"><i class="fa fa-check"></i>Sequence similarity</a>
<ul>
<li class="chapter" data-level="" data-path="module-3---lab-practical.html"><a href="module-3---lab-practical.html#diamond"><i class="fa fa-check"></i>Diamond</a></li>
<li class="chapter" data-level="" data-path="module-3---lab-practical.html"><a href="module-3---lab-practical.html#bakta"><i class="fa fa-check"></i>Bakta</a></li>
<li class="chapter" data-level="" data-path="module-3---lab-practical.html"><a href="module-3---lab-practical.html#resistance-gene-identifier-rgi"><i class="fa fa-check"></i>Resistance Gene Identifier (RGI)</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-3---lab-practical.html"><a href="module-3---lab-practical.html#machine-learning"><i class="fa fa-check"></i>Machine learning</a>
<ul>
<li class="chapter" data-level="" data-path="module-3---lab-practical.html"><a href="module-3---lab-practical.html#kofamscan"><i class="fa fa-check"></i>Kofamscan</a></li>
<li class="chapter" data-level="" data-path="module-3---lab-practical.html"><a href="module-3---lab-practical.html#deepec"><i class="fa fa-check"></i>deepEC</a></li>
<li class="chapter" data-level="" data-path="module-3---lab-practical.html"><a href="module-3---lab-practical.html#deepfri"><i class="fa fa-check"></i>deepFRI</a></li>
<li class="chapter" data-level="" data-path="module-3---lab-practical.html"><a href="module-3---lab-practical.html#proteinbert"><i class="fa fa-check"></i>proteinBERT</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-3---lab-practical.html"><a href="module-3---lab-practical.html#gapseq"><i class="fa fa-check"></i>gapseq</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-4-visualization-and-finding-functional-significance.html"><a href="module-4-visualization-and-finding-functional-significance.html"><i class="fa fa-check"></i>Module 4: Visualization and Finding Functional Significance</a></li>
<li class="chapter" data-level="" data-path="module-4---lab-practical.html"><a href="module-4---lab-practical.html"><i class="fa fa-check"></i>Module 4 - Lab practical</a>
<ul>
<li class="chapter" data-level="" data-path="module-4---lab-practical.html"><a href="module-4---lab-practical.html#setting-up-the-workspace-1"><i class="fa fa-check"></i>Setting up the workspace</a>
<ul>
<li class="chapter" data-level="" data-path="module-4---lab-practical.html"><a href="module-4---lab-practical.html#ssh-port-forwarding-1"><i class="fa fa-check"></i>SSH port forwarding</a></li>
<li class="chapter" data-level="" data-path="module-4---lab-practical.html"><a href="module-4---lab-practical.html#workspace-1"><i class="fa fa-check"></i>Workspace</a></li>
<li class="chapter" data-level="" data-path="module-4---lab-practical.html"><a href="module-4---lab-practical.html#start-container-1"><i class="fa fa-check"></i>Start container</a></li>
<li class="chapter" data-level="" data-path="module-4---lab-practical.html"><a href="module-4---lab-practical.html#jupyter-1"><i class="fa fa-check"></i>Jupyter</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-4---lab-practical.html"><a href="module-4---lab-practical.html#overview-1"><i class="fa fa-check"></i>Overview</a></li>
<li class="chapter" data-level="" data-path="module-4---lab-practical.html"><a href="module-4---lab-practical.html#salmon"><i class="fa fa-check"></i>Salmon</a></li>
<li class="chapter" data-level="" data-path="module-4---lab-practical.html"><a href="module-4---lab-practical.html#deseq2"><i class="fa fa-check"></i>DESeq2</a></li>
<li class="chapter" data-level="" data-path="module-4---lab-practical.html"><a href="module-4---lab-practical.html#gene-set-enrichment-analysis-gsea"><i class="fa fa-check"></i>Gene set enrichment analysis (GSEA)</a></li>
<li class="chapter" data-level="" data-path="module-4---lab-practical.html"><a href="module-4---lab-practical.html#cytoscape"><i class="fa fa-check"></i>Cytoscape</a></li>
</ul></li>
<li class="divider"></li>
<h1 id="sponsors">Sponsors</h1>
 <center>
 <a href="https://www.cihr-irsc.gc.ca"><img src="img/sponsors/cihr.png" width=90% style="padding-bottom: 20px" alt="Canadian Institutes of Health Research logo."></a>

 <a href="https://genomecanada.ca/"><img src="img/sponsors/genomecanada.webp" width=40% style="display:inline-block;"  style="padding-bottom: 20px" alt="Genome Canada logo.">
 <a href="https://www.ontariogenomics.ca/"><img src="img/sponsors/ontariogenomics.png" width=40% style="display:inline-block;" style="padding-bottom: 20px" alt="Ontario Genomics logo.">
 <br><br>
 <a href="https://oicr.on.ca/"><img src="img/sponsors/oicr.png" width=25% style="display:inline-block;" style="padding-bottom: 20px" alt="Ontario Institute for Cancer Research logo.">
 <a href="https://www.hpcforhealth.ca/"><img src="img/sponsors/hpc4health.png" width=60% style="display:inline-block;" style="padding-bottom: 20px "alt="HPC4Health logo.">
 <br><br>
 <li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Advanced Microbiome Analysis 2025</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="module-3---lab-practical" class="section level1 hasAnchor">
<h1>Module 3 - Lab practical<a href="module-3---lab-practical.html#module-3---lab-practical" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>In this lab, we will predict and briefly examine the functions encoded in the assembled metagenome from the previous labs.
Optional context and questions are provided as collapsible sections like so:</p>
<details>
<summary>
Click me
</summary>
<p>Optional context is provided to make this workshop as inclusive as possible. Feel free to skip
any topics that you are already familiar with. Questions and answers will also use this collapsible format.</p>
<hr />
</details>
<p>All instructions required to completed this lab are provided below, so you are free to work asynchronously at your own pace.
Active discussions with both peers and instructors are encouraged. Say hi to the person next to you!</p>
<div id="setting-up-the-workspace" class="section level2 hasAnchor">
<h2>Setting up the workspace<a href="module-3---lab-practical.html#setting-up-the-workspace" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Before we begin, this section will guide you through the set up of your workspace and compute environment.
The software tools used in this lab all have dependencies that must be installed separately.
Constructing the environment for each tool is non-trivial and error-prone.
The difficulty of replicating compute environments for bioinformatics software is one of the key contributing factors
to the reproducibility crisis [1].</p>
<p>We will instead be using prepackaged compute environments called “containers”, implemented by Apptainer (formerly Singularity) [2].
One of these containers will enable the use of jupyter notebooks along with a collection of utilities.</p>
<ol style="list-style-type: decimal">
<li>Mangul S, Martin LS, Eskin E, Blekhman R. Improving the usability and archival stability of bioinformatics software. Genome Biol. 2019;20(1):47. <a href="https://doi.org/10.1186/s13059-019-1649-8" class="uri">https://doi.org/10.1186/s13059-019-1649-8</a></li>
<li>Kurtzer GM, Sochat V, Bauer MW. Singularity: Scientific containers for mobility of compute. PLOS ONE. 2017;12(5):e0177459. <a href="https://doi.org/10.1371/journal.pone.0177459" class="uri">https://doi.org/10.1371/journal.pone.0177459</a></li>
</ol>
<div id="steps-for-setup" class="section level4 hasAnchor">
<h4>Steps for setup:<a href="module-3---lab-practical.html#steps-for-setup" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><img src="img/lab3/compute_env.png" /></p>
<ol style="list-style-type: decimal">
<li>Open an additional port via SSH for the new jupyter notebook</li>
<li>Create a new workspace directory for this lab</li>
<li>Unpack and start the container</li>
<li>Open the jupyter notebook in your browser</li>
</ol>
</div>
<div id="ssh-port-forwarding" class="section level3 hasAnchor">
<h3>SSH port forwarding<a href="module-3---lab-practical.html#ssh-port-forwarding" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We will use SSH port forwarding to open the additional port.</p>
<div id="terminal-unixmacos" class="section level4 hasAnchor">
<h4>Terminal (Unix/MacOS)<a href="module-3---lab-practical.html#terminal-unixmacos" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Add <code>-L 48888:localhost:48888</code> to the SSH command you use to connect to the server. For example:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode numberSource bash numberLines"><code class="sourceCode bash"><span id="cb1-1"><a href="module-3---lab-practical.html#cb1-1"></a><span class="fu">ssh</span> <span class="at">-L</span> 48888:localhost:48888 ubuntu@<span class="va">${your_number}</span>.uhn-hpc.ca <span class="at">-i</span> <span class="va">${path_to_your_key}</span></span></code></pre></div>
<p>This means port <code>48888</code> on the remote server will be forwarded to your machine at <code>localhost:48888</code>.</p>
</div>
<div id="putty-windows" class="section level4 hasAnchor">
<h4>Putty (Windows)<a href="module-3---lab-practical.html#putty-windows" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Modern versions of Windows should have <code>ssh</code> available through Powershell. Putty is available as a backup.</p>
<p><img src="img/lab3/putty_port_forward.png" /></p>
</div>
</div>
<div id="workspace" class="section level3 hasAnchor">
<h3>Workspace<a href="module-3---lab-practical.html#workspace" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Create a new workspace in your home directory.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="module-3---lab-practical.html#cb2-1" tabindex="-1"></a><span class="bu">cd</span> ~</span>
<span id="cb2-2"><a href="module-3---lab-practical.html#cb2-2" tabindex="-1"></a><span class="fu">mkdir</span> amb_lab3</span>
<span id="cb2-3"><a href="module-3---lab-practical.html#cb2-3" tabindex="-1"></a><span class="bu">cd</span> amb_lab3</span></code></pre></div>
<p>Link the resource folder for easy access.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="module-3---lab-practical.html#cb3-1" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> ~/CourseData/MIC_data/amb_module3/ ./lib</span></code></pre></div>
</div>
<div id="start-container" class="section level3 hasAnchor">
<h3>Start container<a href="module-3---lab-practical.html#start-container" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Most containers for this lab were obtained from the biocontainers project, which provides a large collection of bioinformatics software in a standard format.
For each tool, we will provide URIs for the the specific container image used and where to get future versions.
The <code>pull</code> command of <code>apptainer</code> can be used to download images (docs: <a href="https://apptainer.org/docs/user/main/cli/apptainer_pull.html" class="uri">https://apptainer.org/docs/user/main/cli/apptainer_pull.html</a>).</p>
<p>Here is an example which downloads <code>prodigal</code> as the container image <code>example.sif</code> from <code>docker://quay.io/biocontainers/prodigal:2.6.3--h7b50bb2_10</code>.
The image URI takes the form of <code>&lt;protocol&gt;://&lt;address&gt;:&lt;tag&gt;</code>. Tags are used to identify versions <a href="https://quay.io/repository/biocontainers/prodigal?tab=tags" class="uri">https://quay.io/repository/biocontainers/prodigal?tab=tags</a>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb4-1"><a href="module-3---lab-practical.html#cb4-1"></a>apptainer pull example.sif docker:<span class="op">//</span>quay.io<span class="op">/</span>biocontainers<span class="op">/</span>prodigal:<span class="fl">2.6.3</span><span class="op">--</span>h7b50bb2_10</span></code></pre></div>
<p>We can now run the prodigal executable inside the container using <code>exec</code> (execute) like so.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="module-3---lab-practical.html#cb5-1" tabindex="-1"></a><span class="ex">apptainer</span> exec ./example.sif prodigal <span class="at">-h</span></span></code></pre></div>
<p>All containers for this lab were pre-downloaded in the resource folder now linked as <code>./lib</code>, including a main container with Jupyter and miscellaneous utilities.
The main container provides an automated setup script named <code>unpack</code>. Let’s run it.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="module-3---lab-practical.html#cb6-1" tabindex="-1"></a><span class="ex">apptainer</span> exec ./lib/amb_lab3.sif unpack</span></code></pre></div>
<p>The following will dedicate the current terminal to running the juptyer notebook on port 48888.
You may want to connect another terminal to maintain terminal access.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="module-3---lab-practical.html#cb7-1" tabindex="-1"></a><span class="ex">./start_lab</span></span></code></pre></div>
</div>
<div id="jupyter" class="section level3 hasAnchor">
<h3>Jupyter<a href="module-3---lab-practical.html#jupyter" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>JupyterLab should now be accessible at <a href="http://localhost:48888/lab" class="uri">http://localhost:48888/lab</a>. Create a new python notebook and import dependencies.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb8-1"><a href="module-3---lab-practical.html#cb8-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb8-2"><a href="module-3---lab-practical.html#cb8-2"></a><span class="im">from</span> pathlib <span class="im">import</span> Path</span>
<span id="cb8-3"><a href="module-3---lab-practical.html#cb8-3"></a><span class="im">from</span> local.ipc <span class="im">import</span> Shell</span>
<span id="cb8-4"><a href="module-3---lab-practical.html#cb8-4"></a></span>
<span id="cb8-5"><a href="module-3---lab-practical.html#cb8-5"></a>LIB <span class="op">=</span> Path(<span class="st">&quot;./lib&quot;</span>)</span>
<span id="cb8-6"><a href="module-3---lab-practical.html#cb8-6"></a>CPUS <span class="op">=</span> <span class="dv">4</span> <span class="co"># your workshop instance has been alloted 4 cores</span></span></code></pre></div>
<ul>
<li><code>pandas</code> enables parsing and reading of tables</li>
<li><code>pathlib</code> enables manipulation of filesystem paths</li>
<li>Pre-downloaded container images are in <code>LIB</code></li>
<li>Pre-computed outputs for this lab are available at <code>LIB/"outputs"</code>.</li>
</ul>
<p>Since containers can not be run inside other containers,
the provided <code>Shell</code> function grants access to the terminal outside of the main container.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb9-1"><a href="module-3---lab-practical.html#cb9-1"></a>Shell(<span class="ss">f&quot;&quot;&quot;</span></span>
<span id="cb9-2"><a href="module-3---lab-practical.html#cb9-2"></a><span class="ss">pwd -P</span></span>
<span id="cb9-3"><a href="module-3---lab-practical.html#cb9-3"></a><span class="ss">echo </span><span class="sc">{</span>LIB<span class="sc">}</span></span>
<span id="cb9-4"><a href="module-3---lab-practical.html#cb9-4"></a><span class="ss">&quot;&quot;&quot;</span>)</span></code></pre></div>
<details>
<summary>
Python comments <code># ...</code>
</summary>
<p>Comments are free text not meant to be interpreted as python code and are indicated by a preceeding <code>#</code>.
They are useful for leaving notes.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb10-1"><a href="module-3---lab-practical.html#cb10-1"></a><span class="co"># &lt;helpful information&gt;</span></span></code></pre></div>
</details>
<details>
<summary>
Python variables <code>CPUS = 4</code>
</summary>
<p>Useful values can be saved to variables using <code>=</code> for later use.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb11-1"><a href="module-3---lab-practical.html#cb11-1"></a>x <span class="op">=</span> <span class="dv">1</span> <span class="co"># the &quot;value&quot; 1 is assigned to the &quot;variable&quot; x</span></span></code></pre></div>
</details>
<details>
<summary>
Python functions <code>Shell(...)</code>
</summary>
<p>Functions are saved procedures that can be repeatedly executed through “calls”.
Some functions recieve additional information as “arguments” or “return” information back when called.
Called functions execute their procedure and are “evaluated” to their return value.
Functions without a return evaluate to the dedicated null value <code>None</code>.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb12-1"><a href="module-3---lab-practical.html#cb12-1"></a><span class="co"># function definition syntax</span></span>
<span id="cb12-2"><a href="module-3---lab-practical.html#cb12-2"></a><span class="kw">def</span> function_name(): <span class="co"># no arguments</span></span>
<span id="cb12-3"><a href="module-3---lab-practical.html#cb12-3"></a>    <span class="cf">pass</span> <span class="co"># do nothing, return nothing</span></span>
<span id="cb12-4"><a href="module-3---lab-practical.html#cb12-4"></a></span>
<span id="cb12-5"><a href="module-3---lab-practical.html#cb12-5"></a><span class="co"># define the function &quot;add&quot;</span></span>
<span id="cb12-6"><a href="module-3---lab-practical.html#cb12-6"></a><span class="kw">def</span> add(a, b):</span>
<span id="cb12-7"><a href="module-3---lab-practical.html#cb12-7"></a>    <span class="cf">return</span> a <span class="op">+</span> b</span>
<span id="cb12-8"><a href="module-3---lab-practical.html#cb12-8"></a></span>
<span id="cb12-9"><a href="module-3---lab-practical.html#cb12-9"></a><span class="co"># call function &quot;add&quot;</span></span>
<span id="cb12-10"><a href="module-3---lab-practical.html#cb12-10"></a>x <span class="op">=</span> add(<span class="dv">1</span>, <span class="dv">2</span>) <span class="co"># add evaluates to 1 + 2, then to 3, then is assigned to x</span></span>
<span id="cb12-11"><a href="module-3---lab-practical.html#cb12-11"></a>x <span class="co"># x is 3</span></span></code></pre></div>
<p><code>Shell</code> accepts text and executes it as a bash command.</p>
</details>
<details>
<summary>
Python <code>import</code> and dot notation <code>local.ipc</code>
</summary>
<div class="sourceCode" id="cb13"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb13-1"><a href="module-3---lab-practical.html#cb13-1"></a><span class="im">from</span> local.ipc <span class="im">import</span> Shell <span class="co"># access &quot;ipc&quot; from &quot;local&quot;</span></span></code></pre></div>
<p><code>local</code> is a folder containing the file <code>ipc.py</code>, which in turn defines the function <code>Shell</code>.
<code>local</code> is a python “module” and <code>ipc.py</code> is a python “script”.
<code>import</code> makes python components available from other files, typically called “libraries”.
Dot notation (the period between <code>local</code> and <code>ipc</code>) retrieves the named components of complex objects.
An alternative syntax to the above is:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb14-1"><a href="module-3---lab-practical.html#cb14-1"></a><span class="im">import</span> local.ipc.Shell <span class="co"># retrieve Shell from ipc in local</span></span></code></pre></div>
<p>imported components can be renamed using the <code>as</code> keyword:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb15-1"><a href="module-3---lab-practical.html#cb15-1"></a><span class="im">import</span> local.ipc.Shell <span class="im">as</span> Terminal <span class="co"># Shell is now &quot;Terminal&quot;</span></span></code></pre></div>
</details>
<details>
<summary>
Python strings <code>f"..."</code>
</summary>
<p>Text values in python (strings), such as the contents of a bash command, must be declared in quotes like <code>"text"</code> or <code>'text'</code>.
Triple quotes allows multi-line strings. The <code>f</code> prefix enables the evaluation of python variables in the string.
For example, <code>f"echo {LIB}"</code> evaluates to <code>echo ./lib</code>.
This is a convient way to simplify commands by saving repeated sections to variables.</p>
</details>
</div>
</div>
<div id="overview" class="section level2 hasAnchor">
<h2>Overview<a href="module-3---lab-practical.html#overview" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This lab will guide you through finding and identifying the functional units
(represented by the red and orange features below) of a metagenomic assembly, with a focus on proteins and enzymes.</p>
<p><img src="img/lab3/data_layout.png" /></p>
<p>We will begin by exploring techniques related to sequence similarity, which seeks to identify known genes through comparison with reference databases.</p>
<ul>
<li><a href="module-3---lab-practical.html#prodigal">Prodigal</a></li>
<li><a href="module-3---lab-practical.html#diamond">Diamond</a></li>
<li><a href="module-3---lab-practical.html#bakta">Bakta</a></li>
<li><a href="module-3---lab-practical.html#resistance-gene-identifier-rgi">Resistance Gene Identifier (RGI)</a></li>
</ul>
<p>Then we will transition to machine-learning tools, which move away from the “lookup table” approach of sequence similarity.</p>
<ul>
<li><a href="module-3---lab-practical.html#kofamscan">Kofamscan</a></li>
<li><a href="module-3---lab-practical.html#deepfri">deepFRI</a></li>
<li><a href="module-3---lab-practical.html#deepec">deepEC</a></li>
<li><a href="module-3---lab-practical.html#proteinbert">proteinBERT</a></li>
</ul>
<p>Finally, we will use gapseq to construct toy metabolic models using techniques to predict the existence of enzymes.</p>
<ul>
<li><a href="module-3---lab-practical.html#gapseq">gapseq</a></li>
</ul>
</div>
<div id="prodigal" class="section level2 hasAnchor">
<h2>Prodigal<a href="module-3---lab-practical.html#prodigal" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The container image for prodigal:</p>
<p><a href="https://quay.io/repository/biocontainers/prodigal?tab=tags" class="uri">https://quay.io/repository/biocontainers/prodigal?tab=tags</a></p>
<div class="sourceCode" id="cb16"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb16-1"><a href="module-3---lab-practical.html#cb16-1"></a><span class="co"># apptainer pull ./lib/prodigal.sif docker://quay.io/biocontainers/prodigal:2.6.3--h7b50bb2_10</span></span>
<span id="cb16-2"><a href="module-3---lab-practical.html#cb16-2"></a>prodigal_sif <span class="op">=</span> LIB<span class="op">/</span><span class="st">&quot;prodigal.sif&quot;</span></span></code></pre></div>
<p>The goal of prodigal is to predict the protein coding features (open reading frames; ORFs) in a given nucleotide sequence.
Recall that this corresponds to the red features in the overview diagram, neglecting the orange features for now.
It does this by optimizing predictions based on a set of tuned rules (heuristics) which take into account
gene length and overlap, start and stop codons, ribosome binding sites, and GC content [1].</p>
<ol style="list-style-type: decimal">
<li>Hyatt D, Chen GL, LoCascio PF, Land ML, Larimer FW, Hauser LJ. Prodigal: prokaryotic gene recognition and translation initiation site identification. BMC Bioinformatics. 2010;11(1):119. <a href="https://doi.org/10.1186/1471-2105-11-119" class="uri">https://doi.org/10.1186/1471-2105-11-119</a></li>
</ol>
<p>Based on this, we would expect prodigal to:</p>
<ul>
<li>accept a nucleotide sequence as input</li>
<li>output an itemized list of predicted open reading frames</li>
</ul>
<p>The input sequence will be the metagenomic assembly since we would otherwise neglect some contigs that were not assigned to MAGs,
thereby causing information loss.
Let’s have a look at prodigal’s help message to get an idea of specific inputs and outputs.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb17-1"><a href="module-3---lab-practical.html#cb17-1"></a>Shell(<span class="ss">f&quot;apptainer exec </span><span class="sc">{</span>prodigal_sif<span class="sc">}</span><span class="ss"> prodigal -h&quot;</span>)</span></code></pre></div>
<p>This looks like the input:</p>
<ul>
<li><code>-i:  Specify FASTA/Genbank input file (default reads from stdin).</code></li>
</ul>
<p>These look like they are useful for the output:</p>
<ul>
<li><code>-a:  Write protein translations to the selected file.</code></li>
<li><code>-d:  Write nucleotide sequences of genes to the selected file.</code> These will be helpful next lab for aligning transcripts to genes.</li>
<li><code>-f:  Select output format (gbk, gff, or sco).  Default is gbk.</code> We will choose <code>gff</code> since the genbank output of prodigal is not in a standard formatted.</li>
<li><code>-o:  Specify output file (default writes to stdout).</code></li>
</ul>
<p>After a bit of investigation, it also looks like we should specify that the input is a metagenomic assembly.</p>
<ul>
<li><code>-p:  Select procedure (single or meta).  Default is single.</code></li>
</ul>
<p>Have a go at compiling the final command.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb18-1"><a href="module-3---lab-practical.html#cb18-1"></a>out_dir <span class="op">=</span> <span class="st">&quot;./prodigal&quot;</span></span>
<span id="cb18-2"><a href="module-3---lab-practical.html#cb18-2"></a>Shell(<span class="ss">f&quot;&quot;&quot;</span></span>
<span id="cb18-3"><a href="module-3---lab-practical.html#cb18-3"></a><span class="ss">mkdir -p </span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/</span></span>
<span id="cb18-4"><a href="module-3---lab-practical.html#cb18-4"></a><span class="ss">apptainer exec </span><span class="sc">{</span>prodigal_sif<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb18-5"><a href="module-3---lab-practical.html#cb18-5"></a><span class="ss">    prodigal </span><span class="op">\</span></span>
<span id="cb18-6"><a href="module-3---lab-practical.html#cb18-6"></a><span class="ss">        -i </span><span class="sc">{</span>LIB<span class="sc">}</span><span class="ss">/inputs/assembly.fa </span><span class="op">\</span></span>
<span id="cb18-7"><a href="module-3---lab-practical.html#cb18-7"></a><span class="ss">        ...</span></span>
<span id="cb18-8"><a href="module-3---lab-practical.html#cb18-8"></a><span class="ss">&quot;&quot;&quot;</span>)</span></code></pre></div>
<details>
<summary>
The purpose of <code>\</code>
</summary>
Starting from <code>apptainer exec</code>, the command must be written in a single line.
The backslash <code>\</code> is used to ignore the newline character and continue the command on the next line.
This let’s us write break up a long command into multiple lines for readability.
—
</details>
<details>
<summary>
Suggested answer
</summary>
<div class="sourceCode" id="cb19"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb19-1"><a href="module-3---lab-practical.html#cb19-1"></a>out_dir <span class="op">=</span> Path(<span class="st">&quot;./prodigal&quot;</span>)</span>
<span id="cb19-2"><a href="module-3---lab-practical.html#cb19-2"></a>prodigal_aa <span class="op">=</span> out_dir<span class="op">/</span><span class="st">&quot;orfs.faa&quot;</span></span>
<span id="cb19-3"><a href="module-3---lab-practical.html#cb19-3"></a>prodigal_gff <span class="op">=</span> out_dir<span class="op">/</span><span class="st">&quot;orfs.gff&quot;</span></span>
<span id="cb19-4"><a href="module-3---lab-practical.html#cb19-4"></a>Shell(<span class="ss">f&quot;&quot;&quot;</span><span class="op">\</span></span>
<span id="cb19-5"><a href="module-3---lab-practical.html#cb19-5"></a><span class="ss">mkdir -p </span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/</span></span>
<span id="cb19-6"><a href="module-3---lab-practical.html#cb19-6"></a><span class="ss">apptainer exec </span><span class="sc">{</span>prodigal_sif<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb19-7"><a href="module-3---lab-practical.html#cb19-7"></a><span class="ss">    prodigal </span><span class="op">\</span></span>
<span id="cb19-8"><a href="module-3---lab-practical.html#cb19-8"></a><span class="ss">        -p meta </span><span class="op">\</span></span>
<span id="cb19-9"><a href="module-3---lab-practical.html#cb19-9"></a><span class="ss">        -i </span><span class="sc">{</span>LIB<span class="sc">}</span><span class="ss">/inputs/assembly.fa </span><span class="op">\</span></span>
<span id="cb19-10"><a href="module-3---lab-practical.html#cb19-10"></a><span class="ss">        -a </span><span class="sc">{</span>prodigal_aa<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb19-11"><a href="module-3---lab-practical.html#cb19-11"></a><span class="ss">        -d </span><span class="sc">{</span>prodigal_aa<span class="sc">.</span>with_suffix(<span class="st">&#39;.fna&#39;</span>)<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb19-12"><a href="module-3---lab-practical.html#cb19-12"></a><span class="ss">        -f gff </span><span class="op">\</span></span>
<span id="cb19-13"><a href="module-3---lab-practical.html#cb19-13"></a><span class="ss">        -o </span><span class="sc">{</span>prodigal_gff<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb19-14"><a href="module-3---lab-practical.html#cb19-14"></a><span class="ss">&quot;&quot;&quot;</span>)</span></code></pre></div>
<hr />
</details>
<p>Let’s take a peek at the <code>gff</code>.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb20-1"><a href="module-3---lab-practical.html#cb20-1"></a>gff <span class="op">=</span> pd.read_csv(out_dir<span class="op">/</span><span class="st">&quot;orfs.gff&quot;</span>, sep<span class="op">=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, header<span class="op">=</span><span class="va">None</span>, comment<span class="op">=</span><span class="st">&quot;#&quot;</span>)</span>
<span id="cb20-2"><a href="module-3---lab-practical.html#cb20-2"></a><span class="bu">print</span>(gff.shape)</span>
<span id="cb20-3"><a href="module-3---lab-practical.html#cb20-3"></a>gff.head(<span class="dv">2</span>)</span></code></pre></div>
<p>From left to right, the columns are:</p>
<ol start="0" style="list-style-type: decimal">
<li>contig id</li>
<li>name of the program that generated this feature</li>
<li>feature type, where <code>CDS</code> stands for coding sequence</li>
<li>start position (first nucleotide is 1)</li>
<li>end position</li>
<li>raw prodigal score</li>
<li>strand, where <code>+</code> is forward and <code>-</code> is reverse</li>
<li>codon frame, being one of 0, 1 or 2</li>
<li>a semicolon-separated list of tag-value pairs, providing additional information about each feature</li>
</ol>
<details>
<summary>
Selecting entries in python dataframes and matrices
</summary>
<p>Entries (rows, columns, or cells) in pandas dataframes and numpy matricies can be selected in multiple ways.</p>
<p>Example matrix:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb21-1"><a href="module-3---lab-practical.html#cb21-1"></a>mat <span class="op">=</span> np.array([<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>])</span></code></pre></div>
<p>Select by single index, where negative indexes count from the end:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb22-1"><a href="module-3---lab-practical.html#cb22-1"></a>mat[<span class="dv">0</span>]      <span class="co"># 1</span></span>
<span id="cb22-2"><a href="module-3---lab-practical.html#cb22-2"></a>mat[<span class="op">-</span><span class="dv">1</span>]     <span class="co"># 5</span></span></code></pre></div>
<p>a slice in the form of <code>start:end</code>, where <code>start</code> is inclusive and <code>end</code> is exclusive:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb23-1"><a href="module-3---lab-practical.html#cb23-1"></a>mat[<span class="dv">1</span>:<span class="dv">4</span>]    <span class="co"># [2, 3, 4]</span></span>
<span id="cb23-2"><a href="module-3---lab-practical.html#cb23-2"></a>mat[:<span class="dv">2</span>]     <span class="co"># [1, 2]</span></span>
<span id="cb23-3"><a href="module-3---lab-practical.html#cb23-3"></a>mat[<span class="dv">2</span>:]     <span class="co"># [3, 4, 5]</span></span>
<span id="cb23-4"><a href="module-3---lab-practical.html#cb23-4"></a>mat[<span class="op">-</span><span class="dv">1</span>:]    <span class="co"># [5]</span></span></code></pre></div>
<p>a list of indexes:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb24-1"><a href="module-3---lab-practical.html#cb24-1"></a>mat[[<span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">4</span>]] <span class="co"># [1, 3, 5]</span></span></code></pre></div>
<p>a boolean mask:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb25-1"><a href="module-3---lab-practical.html#cb25-1"></a>mat[[<span class="va">False</span>, <span class="va">False</span>, <span class="va">False</span>, <span class="va">True</span>, <span class="va">True</span>]] <span class="co"># [4, 5]</span></span></code></pre></div>
<hr />
</details>
<p>Let’s investigate how many ORFs were not on MAGs. To do this, we will first create a blacklist for mag contig IDs.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb26-1"><a href="module-3---lab-practical.html#cb26-1"></a>mag_ids <span class="op">=</span> <span class="bu">set</span>()</span>
<span id="cb26-2"><a href="module-3---lab-practical.html#cb26-2"></a><span class="co"># loop through the paths to each of the MAG files</span></span>
<span id="cb26-3"><a href="module-3---lab-practical.html#cb26-3"></a><span class="cf">for</span> mag_file <span class="kw">in</span> [</span>
<span id="cb26-4"><a href="module-3---lab-practical.html#cb26-4"></a>    LIB<span class="op">/</span><span class="st">&quot;inputs/HMP2_MAG_00001-contigs.fa&quot;</span>,</span>
<span id="cb26-5"><a href="module-3---lab-practical.html#cb26-5"></a>    LIB<span class="op">/</span><span class="st">&quot;inputs/HMP2_MAG_00002-contigs.fa&quot;</span>,</span>
<span id="cb26-6"><a href="module-3---lab-practical.html#cb26-6"></a>]:</span>
<span id="cb26-7"><a href="module-3---lab-practical.html#cb26-7"></a>    <span class="co"># we can use SeqIO to iterate through each sequence in a fasta file</span></span>
<span id="cb26-8"><a href="module-3---lab-practical.html#cb26-8"></a>    <span class="cf">for</span> contig <span class="kw">in</span> SeqIO.parse(mag_file, <span class="st">&quot;fasta&quot;</span>):</span>
<span id="cb26-9"><a href="module-3---lab-practical.html#cb26-9"></a>        mag_ids.add(contig.<span class="bu">id</span>) <span class="co"># add the contig id of each contig to mag_ids</span></span>
<span id="cb26-10"><a href="module-3---lab-practical.html#cb26-10"></a><span class="bu">len</span>(mag_ids)</span></code></pre></div>
<details>
<summary>
Python <code>for</code> loops
</summary>
<p>Loops enable repeated execution of a block of code for each element in a collection.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb27-1"><a href="module-3---lab-practical.html#cb27-1"></a>collection <span class="op">=</span> [<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>] <span class="co"># a &quot;list&quot; collection</span></span>
<span id="cb27-2"><a href="module-3---lab-practical.html#cb27-2"></a><span class="cf">for</span> element <span class="kw">in</span> collection:</span>
<span id="cb27-3"><a href="module-3---lab-practical.html#cb27-3"></a>    <span class="bu">print</span>(element) <span class="co"># do something with each element</span></span>
<span id="cb27-4"><a href="module-3---lab-practical.html#cb27-4"></a></span>
<span id="cb27-5"><a href="module-3---lab-practical.html#cb27-5"></a><span class="co"># result:</span></span>
<span id="cb27-6"><a href="module-3---lab-practical.html#cb27-6"></a><span class="co"># 1</span></span>
<span id="cb27-7"><a href="module-3---lab-practical.html#cb27-7"></a><span class="co"># 2</span></span>
<span id="cb27-8"><a href="module-3---lab-practical.html#cb27-8"></a><span class="co"># 3</span></span></code></pre></div>
</details>
<p>Then filter the gff based on the MAG blacklist.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb28-1"><a href="module-3---lab-practical.html#cb28-1"></a>_filter <span class="op">=</span> <span class="op">~</span>gff[<span class="dv">0</span>].isin(mag_ids) <span class="co"># column 0 is the contig id</span></span>
<span id="cb28-2"><a href="module-3---lab-practical.html#cb28-2"></a>non_mag_orfs <span class="op">=</span> gff[_filter]</span>
<span id="cb28-3"><a href="module-3---lab-practical.html#cb28-3"></a>non_mag_orfs.shape</span></code></pre></div>
<details>
<summary>
How many ORFs were not from MAGs?
</summary>
1060 of 7176, or about 15%
</details>
</div>
<div id="sequence-similarity" class="section level2 hasAnchor">
<h2>Sequence similarity<a href="module-3---lab-practical.html#sequence-similarity" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this section, we will use tools to find proteins with known sequences that are similar to that of the predicted ORFs.
By assuming that similar sequences equate to similar functions, we can assign functions through reference databases of known proteins.
While not perfect [1, 2], this is a common first-pass approach to assign functions to genes in metagenomic assemblies.</p>
<ol style="list-style-type: decimal">
<li>Omelchenko MV, Galperin MY, Wolf YI, Koonin EV. Non-homologous isofunctional enzymes: A systematic analysis of alternative solutions in enzyme evolution. Biol Direct. 2010;5(1):31. <a href="https://doi.org/10.1186/1745-6150-5-31" class="uri">https://doi.org/10.1186/1745-6150-5-31</a></li>
<li>Seffernick JL, de Souza ML, Sadowsky MJ, Wackett LP. Melamine deaminase and atrazine chlorohydrolase: 98 percent identical but functionally different. J Bacteriol. 2001;183(8):2405–10. <a href="https://doi.org/10.1128/JB.183.8.2405-2410.2001" class="uri">https://doi.org/10.1128/JB.183.8.2405-2410.2001</a></li>
</ol>
<div id="diamond" class="section level3 hasAnchor">
<h3>Diamond<a href="module-3---lab-practical.html#diamond" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><a href="https://quay.io/repository/biocontainers/diamond?tab=tags" class="uri">https://quay.io/repository/biocontainers/diamond?tab=tags</a></p>
<div class="sourceCode" id="cb29"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb29-1"><a href="module-3---lab-practical.html#cb29-1"></a><span class="co"># docker://quay.io/biocontainers/diamond:2.1.11--h5ca1c30_2</span></span>
<span id="cb29-2"><a href="module-3---lab-practical.html#cb29-2"></a>diamond_sif <span class="op">=</span> LIB<span class="op">/</span><span class="st">&quot;diamond.sif&quot;</span></span></code></pre></div>
<p>Diamond is highly optimized for finding similar protein sequences between a list of queries and a list of subjects [1].
It has a similar command line interface (CLI) to the classic BLAST, requiring that we first build a database from the list of subjects,
before searching against it with a list of queries.</p>
<ol style="list-style-type: decimal">
<li>Buchfink B, Xie C, Huson DH. Fast and sensitive protein alignment using DIAMOND. Nat Methods. 2015;12(1):59–60. <a href="https://doi.org/10.1038/nmeth.3176" class="uri">https://doi.org/10.1038/nmeth.3176</a></li>
</ol>
<p>We expect to run diamond in two steps:</p>
<ol style="list-style-type: decimal">
<li>Build a reference database
<ul>
<li>inputs: a list of protein sequences</li>
<li>outputs: a database file</li>
</ul></li>
<li>Search the database
<ul>
<li>inputs: a list of protein sequences, the compiled database</li>
<li>outputs: a list of hits</li>
</ul></li>
</ol>
<div class="sourceCode" id="cb30"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb30-1"><a href="module-3---lab-practical.html#cb30-1"></a>Shell(<span class="ss">f&quot;apptainer exec </span><span class="sc">{</span>diamond_sif<span class="sc">}</span><span class="ss"> diamond --help&quot;</span>)</span></code></pre></div>
<p>We will build a database for transporters using the transporter classification database (TCDB) [1].
TCDB has been predownloaded from these URLs.</p>
<ul>
<li>protein sequences: <a href="https://www.tcdb.org/public/tcdb" class="uri">https://www.tcdb.org/public/tcdb</a></li>
<li>sequences are grouped into families: <a href="https://www.tcdb.org/cgi-bin/projectv/public/families.py" class="uri">https://www.tcdb.org/cgi-bin/projectv/public/families.py</a></li>
<li>transporter substrates: <a href="https://www.tcdb.org/cgi-bin/substrates/getSubstrates.py" class="uri">https://www.tcdb.org/cgi-bin/substrates/getSubstrates.py</a></li>
</ul>
<ol style="list-style-type: decimal">
<li>Saier MH Jr, Tran CV, Barabote RD. TCDB: the Transporter Classification Database for membrane transport protein analyses and information. Nucleic Acids Res. 2006;34(suppl_1):D181–6. <a href="https://doi.org/10.1093/nar/gkj001" class="uri">https://doi.org/10.1093/nar/gkj001</a></li>
</ol>
<div class="sourceCode" id="cb31"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb31-1"><a href="module-3---lab-practical.html#cb31-1"></a>Shell(<span class="ss">f&quot;apptainer exec </span><span class="sc">{</span>diamond_sif<span class="sc">}</span><span class="ss"> diamond makedb&quot;</span>)</span></code></pre></div>
<p>The important arguments appear to be:</p>
<ul>
<li><code>--db                     database file</code></li>
<li><code>--in                     input reference file in FASTA format/input DAA files for merge-daa</code></li>
</ul>
<p>Have a go at compiling the command to build the diamond database.
Running a your command is a good way to check for correctness and resulting error messages can be informative.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb32-1"><a href="module-3---lab-practical.html#cb32-1"></a>tcdb_fasta <span class="op">=</span> LIB<span class="op">/</span><span class="st">&quot;transporter_classification_db/tcdb.faa&quot;</span></span>
<span id="cb32-2"><a href="module-3---lab-practical.html#cb32-2"></a>tcdb_db <span class="op">=</span> Path(<span class="st">&quot;./diamond/tcdb.dmnd&quot;</span>)</span>
<span id="cb32-3"><a href="module-3---lab-practical.html#cb32-3"></a>Shell(<span class="ss">f&quot;&quot;&quot;</span></span>
<span id="cb32-4"><a href="module-3---lab-practical.html#cb32-4"></a><span class="ss">apptainer exec </span><span class="sc">{</span>diamond_sif<span class="sc">}</span><span class="ss"> diamond makedb </span><span class="op">\</span></span>
<span id="cb32-5"><a href="module-3---lab-practical.html#cb32-5"></a><span class="ss">    ...</span></span>
<span id="cb32-6"><a href="module-3---lab-practical.html#cb32-6"></a><span class="ss">&quot;&quot;&quot;</span>)</span></code></pre></div>
<details>
<summary>
Possible solution
</summary>
<div class="sourceCode" id="cb33"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb33-1"><a href="module-3---lab-practical.html#cb33-1"></a>tcdb_fasta <span class="op">=</span> LIB<span class="op">/</span><span class="st">&quot;transporter_classification_db/tcdb.faa&quot;</span></span>
<span id="cb33-2"><a href="module-3---lab-practical.html#cb33-2"></a>tcdb_db <span class="op">=</span> Path(<span class="st">&quot;./diamond/tcdb.dmnd&quot;</span>)</span>
<span id="cb33-3"><a href="module-3---lab-practical.html#cb33-3"></a>Shell(<span class="ss">f&quot;&quot;&quot;</span></span>
<span id="cb33-4"><a href="module-3---lab-practical.html#cb33-4"></a><span class="ss">mkdir -p </span><span class="sc">{</span>tcdb_db<span class="sc">.</span>parent<span class="sc">}</span><span class="ss"> # parent folder of given path</span></span>
<span id="cb33-5"><a href="module-3---lab-practical.html#cb33-5"></a><span class="ss">apptainer exec </span><span class="sc">{</span>diamond_sif<span class="sc">}</span><span class="ss"> diamond makedb </span><span class="op">\</span></span>
<span id="cb33-6"><a href="module-3---lab-practical.html#cb33-6"></a><span class="ss">    --threads </span><span class="sc">{</span>CPUS<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb33-7"><a href="module-3---lab-practical.html#cb33-7"></a><span class="ss">    --in </span><span class="sc">{</span>tcdb_fasta<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb33-8"><a href="module-3---lab-practical.html#cb33-8"></a><span class="ss">    --db </span><span class="sc">{</span>tcdb_db<span class="sc">}</span></span>
<span id="cb33-9"><a href="module-3---lab-practical.html#cb33-9"></a><span class="ss">&quot;&quot;&quot;</span>)</span></code></pre></div>
</details>
<p>Now we can search our predicted ORFs for transporters catalogued in TCDB.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb34-1"><a href="module-3---lab-practical.html#cb34-1"></a>Shell(<span class="ss">f&quot;apptainer exec </span><span class="sc">{</span>diamond_sif<span class="sc">}</span><span class="ss"> diamond blastp&quot;</span>)</span></code></pre></div>
<p>That produced a lot of output, but these are the essential arguments:</p>
<ul>
<li><code>--query                  input query file</code></li>
<li><code>--db                     database file</code></li>
<li><code>--out                    output file</code></li>
</ul>
<p>These may be frequently useful:</p>
<ul>
<li><code>--fast                   enable fast mode</code></li>
<li><code>--sensitive              enable sensitive mode</code></li>
<li><code>--ultra-sensitive        enable ultra sensitive mode</code></li>
<li><code>--threads                number of CPU threads</code></li>
<li><code>--outfmt                 output format</code></li>
<li><code>--evalue                 maximum e-value to report alignments (default=0.001)</code></li>
</ul>
<p>I find that this works well for general cases.</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb35-1"><a href="module-3---lab-practical.html#cb35-1"></a>columns <span class="op">=</span> <span class="st">&quot;qseqid sseqid score pident evalue&quot;</span></span>
<span id="cb35-2"><a href="module-3---lab-practical.html#cb35-2"></a>diamond_out <span class="op">=</span> Path(<span class="st">&quot;./diamond/orfs.tsv&quot;</span>)</span>
<span id="cb35-3"><a href="module-3---lab-practical.html#cb35-3"></a>Shell(<span class="ss">f&quot;&quot;&quot;</span></span>
<span id="cb35-4"><a href="module-3---lab-practical.html#cb35-4"></a><span class="ss">apptainer exec </span><span class="sc">{</span>diamond_sif<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb35-5"><a href="module-3---lab-practical.html#cb35-5"></a><span class="ss">    diamond blastp </span><span class="op">\</span></span>
<span id="cb35-6"><a href="module-3---lab-practical.html#cb35-6"></a><span class="ss">        --query ./prodigal/orfs.faa </span><span class="op">\</span></span>
<span id="cb35-7"><a href="module-3---lab-practical.html#cb35-7"></a><span class="ss">        --db </span><span class="sc">{</span>tcdb_db<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb35-8"><a href="module-3---lab-practical.html#cb35-8"></a><span class="ss">        --out </span><span class="sc">{</span>diamond_out<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb35-9"><a href="module-3---lab-practical.html#cb35-9"></a><span class="ss">        --outfmt 6 </span><span class="sc">{</span>columns<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb35-10"><a href="module-3---lab-practical.html#cb35-10"></a><span class="ss">        --threads </span><span class="sc">{</span>CPUS<span class="sc">}</span></span>
<span id="cb35-11"><a href="module-3---lab-practical.html#cb35-11"></a><span class="ss">&quot;&quot;&quot;</span>)</span></code></pre></div>
<p>In addition to e-evalue, we can calculate a blast score ratio (BSR) to assign confidence to the hits.
BSR for a hit is defined as the ratio of its score to the score of the query aligned to itself (the best possible score).
A BSR of less than 0.4 was found to be a good threshold [1].</p>
<ol style="list-style-type: decimal">
<li>Rost B. Twilight zone of protein sequence alignments. Protein Eng. 1999;12(2):85–94. <a href="https://doi.org/10.1093/protein/12.2.85" class="uri">https://doi.org/10.1093/protein/12.2.85</a></li>
</ol>
<p>Perform a self blast. We can skip making a database since the number of sequences is small.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb36-1"><a href="module-3---lab-practical.html#cb36-1"></a>diamond_self <span class="op">=</span> Path(<span class="st">&quot;./diamond/self.raw&quot;</span>)</span>
<span id="cb36-2"><a href="module-3---lab-practical.html#cb36-2"></a>Shell(<span class="ss">f&quot;&quot;&quot;</span></span>
<span id="cb36-3"><a href="module-3---lab-practical.html#cb36-3"></a><span class="ss">apptainer exec </span><span class="sc">{</span>diamond_sif<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb36-4"><a href="module-3---lab-practical.html#cb36-4"></a><span class="ss">    diamond blastp </span><span class="op">\</span></span>
<span id="cb36-5"><a href="module-3---lab-practical.html#cb36-5"></a><span class="ss">        --query </span><span class="sc">{</span>prodigal_aa<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb36-6"><a href="module-3---lab-practical.html#cb36-6"></a><span class="ss">        --db </span><span class="sc">{</span>prodigal_aa<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb36-7"><a href="module-3---lab-practical.html#cb36-7"></a><span class="ss">        --out </span><span class="sc">{</span>diamond_self<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb36-8"><a href="module-3---lab-practical.html#cb36-8"></a><span class="ss">        --outfmt 6 </span><span class="sc">{</span>columns<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb36-9"><a href="module-3---lab-practical.html#cb36-9"></a><span class="ss">        --max-target-seqs 1 </span><span class="op">\</span></span>
<span id="cb36-10"><a href="module-3---lab-practical.html#cb36-10"></a><span class="ss">        --faster </span><span class="op">\</span></span>
<span id="cb36-11"><a href="module-3---lab-practical.html#cb36-11"></a><span class="ss">        --threads </span><span class="sc">{</span>CPUS<span class="sc">}</span></span>
<span id="cb36-12"><a href="module-3---lab-practical.html#cb36-12"></a><span class="ss">&quot;&quot;&quot;</span>)</span></code></pre></div>
<details>
<summary>
“string”.split() and “string”.join()
</summary>
<p>Strings can be split into lists using the <code>split</code> method.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb37-1"><a href="module-3---lab-practical.html#cb37-1"></a><span class="co">&quot;a b c&quot;</span>.split(<span class="st">&quot; &quot;</span>) <span class="co"># [&quot;a&quot;, &quot;b&quot;, &quot;c&quot;]</span></span>
<span id="cb37-2"><a href="module-3---lab-practical.html#cb37-2"></a><span class="co">&quot;a|b|c&quot;</span>.split(<span class="st">&quot;|&quot;</span>) <span class="co"># [&quot;a&quot;, &quot;b&quot;, &quot;c&quot;]</span></span>
<span id="cb37-3"><a href="module-3---lab-practical.html#cb37-3"></a><span class="co">&quot;c_001_3&quot;</span>.split(<span class="st">&quot;_&quot;</span>) <span class="co"># [&quot;c&quot;, &quot;001&quot;, &quot;3&quot;]</span></span></code></pre></div>
<p>Lists can be joined into strings using the <code>join</code> method.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb38-1"><a href="module-3---lab-practical.html#cb38-1"></a><span class="co">&quot; &quot;</span>.join([<span class="st">&quot;a&quot;</span>, <span class="st">&quot;b&quot;</span>, <span class="st">&quot;c&quot;</span>]) <span class="co"># &quot;a b c&quot;</span></span>
<span id="cb38-2"><a href="module-3---lab-practical.html#cb38-2"></a><span class="co">&quot;_&quot;</span>.join([<span class="st">&quot;c&quot;</span>, <span class="st">&quot;001&quot;</span>]) <span class="co"># &quot;c_001&quot;</span></span></code></pre></div>
<hr />
</details>
<p>We now remove hits with a BSR &lt; 0.4. First, load in the self blast results.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb39-1"><a href="module-3---lab-practical.html#cb39-1"></a><span class="co"># get self scores</span></span>
<span id="cb39-2"><a href="module-3---lab-practical.html#cb39-2"></a>df_self <span class="op">=</span> pd.read_csv(diamond_self, sep<span class="op">=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, header<span class="op">=</span><span class="va">None</span>, names<span class="op">=</span>columns.split(<span class="st">&quot; &quot;</span>))</span>
<span id="cb39-3"><a href="module-3---lab-practical.html#cb39-3"></a>df_self <span class="op">=</span> df_self[df_self[<span class="st">&quot;qseqid&quot;</span>] <span class="op">==</span> df_self[<span class="st">&quot;sseqid&quot;</span>]] <span class="co"># use boolean mask to select self hits</span></span>
<span id="cb39-4"><a href="module-3---lab-practical.html#cb39-4"></a><span class="bu">print</span>(df_self.shape)</span>
<span id="cb39-5"><a href="module-3---lab-practical.html#cb39-5"></a>df_self.head(<span class="dv">2</span>)</span></code></pre></div>
<p>As the code gets more complex, it may be helpful to examine individual components to increase understanding.
For example, have a look at:</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb40-1"><a href="module-3---lab-practical.html#cb40-1"></a>columns.split(<span class="st">&quot; &quot;</span>)</span>
<span id="cb40-2"><a href="module-3---lab-practical.html#cb40-2"></a><span class="co"># and</span></span>
<span id="cb40-3"><a href="module-3---lab-practical.html#cb40-3"></a>df_self[<span class="st">&quot;qseqid&quot;</span>] <span class="op">==</span> df_self[<span class="st">&quot;sseqid&quot;</span>]</span>
<span id="cb40-4"><a href="module-3---lab-practical.html#cb40-4"></a><span class="co"># or even</span></span>
<span id="cb40-5"><a href="module-3---lab-practical.html#cb40-5"></a>df_self[<span class="st">&quot;qseqid&quot;</span>]</span></code></pre></div>
<p>Calculate BSR</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb41-1"><a href="module-3---lab-practical.html#cb41-1"></a>df_raw <span class="op">=</span> pd.read_csv(diamond_raw, sep<span class="op">=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, header<span class="op">=</span><span class="va">None</span>, names<span class="op">=</span>columns.split(<span class="st">&quot; &quot;</span>))</span>
<span id="cb41-2"><a href="module-3---lab-practical.html#cb41-2"></a><span class="co"># merge in self scores by matching the qseqid</span></span>
<span id="cb41-3"><a href="module-3---lab-practical.html#cb41-3"></a><span class="co"># &quot;left&quot; refers to a &quot;left join&quot; where</span></span>
<span id="cb41-4"><a href="module-3---lab-practical.html#cb41-4"></a><span class="co"># each row in the left dataframe (df_raw) is found a match in the right (df_self)</span></span>
<span id="cb41-5"><a href="module-3---lab-practical.html#cb41-5"></a><span class="co"># duplicate columns from the right df are suffixed with &quot;_self&quot;, such as &quot;score_self&quot;</span></span>
<span id="cb41-6"><a href="module-3---lab-practical.html#cb41-6"></a>df_hits <span class="op">=</span> df_raw.merge(df_self, on<span class="op">=</span><span class="st">&quot;qseqid&quot;</span>, suffixes<span class="op">=</span>(<span class="st">&quot;&quot;</span>, <span class="st">&quot;_self&quot;</span>), how<span class="op">=</span><span class="st">&quot;left&quot;</span>)</span>
<span id="cb41-7"><a href="module-3---lab-practical.html#cb41-7"></a><span class="co"># calculate bsr and place in new column</span></span>
<span id="cb41-8"><a href="module-3---lab-practical.html#cb41-8"></a>df_hits[<span class="st">&quot;bsr&quot;</span>] <span class="op">=</span> df_hits[<span class="st">&quot;score&quot;</span>] <span class="op">/</span> df_hits[<span class="st">&quot;score_self&quot;</span>] <span class="co"># divide each value in score by the corresponding value in score_self</span></span>
<span id="cb41-9"><a href="module-3---lab-practical.html#cb41-9"></a>df_hits.head(<span class="dv">2</span>)</span></code></pre></div>
<p>Filter out hits with BSR &lt; 0.4</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb42-1"><a href="module-3---lab-practical.html#cb42-1"></a>df_hits <span class="op">=</span> df_hits[df_hits[<span class="st">&quot;bsr&quot;</span>] <span class="op">&gt;=</span> <span class="fl">0.4</span>] <span class="co"># filter out rows with bsr &lt; 0.4</span></span>
<span id="cb42-2"><a href="module-3---lab-practical.html#cb42-2"></a>df_hits <span class="op">=</span> df_hits[columns.split(<span class="st">&quot; &quot;</span>) <span class="op">+</span> [<span class="st">&quot;bsr&quot;</span>]] <span class="co"># keep only the columns we want</span></span>
<span id="cb42-3"><a href="module-3---lab-practical.html#cb42-3"></a>df_hits.head(<span class="dv">2</span>)</span></code></pre></div>
<p>Take the best hit</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb43-1"><a href="module-3---lab-practical.html#cb43-1"></a><span class="co"># sort by bsr and group by qseqid, keeping the first row of each group</span></span>
<span id="cb43-2"><a href="module-3---lab-practical.html#cb43-2"></a><span class="co"># this will keep the best hit for each query sequence</span></span>
<span id="cb43-3"><a href="module-3---lab-practical.html#cb43-3"></a><span class="co"># groupby(&quot;qseqid&quot;) will set the index to qseqid, so we reset the index</span></span>
<span id="cb43-4"><a href="module-3---lab-practical.html#cb43-4"></a>df_hits <span class="op">=</span> df_hits.sort_values(<span class="st">&quot;bsr&quot;</span>, ascending<span class="op">=</span><span class="va">False</span>).groupby(<span class="st">&quot;qseqid&quot;</span>).first().reset_index()</span>
<span id="cb43-5"><a href="module-3---lab-practical.html#cb43-5"></a>df_hits <span class="op">=</span> df_hits.sort_values(<span class="st">&quot;qseqid&quot;</span>) <span class="co"># re-sort by qseqid</span></span>
<span id="cb43-6"><a href="module-3---lab-practical.html#cb43-6"></a><span class="bu">print</span>(df_hits.shape)</span>
<span id="cb43-7"><a href="module-3---lab-practical.html#cb43-7"></a>df_hits.head(<span class="dv">2</span>)</span></code></pre></div>
<p>Let’s plot the results using a simple diverging bar chart for each MAG. First, let’s create whitelists for the contigs in each MAG.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb44-1"><a href="module-3---lab-practical.html#cb44-1"></a>MAG1 <span class="op">=</span> <span class="bu">set</span>()</span>
<span id="cb44-2"><a href="module-3---lab-practical.html#cb44-2"></a><span class="cf">for</span> e <span class="kw">in</span> SeqIO.parse(LIB<span class="op">/</span><span class="st">&quot;inputs/HMP2_MAG_00001-contigs.fa&quot;</span>, <span class="st">&quot;fasta&quot;</span>):</span>
<span id="cb44-3"><a href="module-3---lab-practical.html#cb44-3"></a>    MAG1.add(e.<span class="bu">id</span>)</span>
<span id="cb44-4"><a href="module-3---lab-practical.html#cb44-4"></a>MAG2 <span class="op">=</span> <span class="bu">set</span>()</span>
<span id="cb44-5"><a href="module-3---lab-practical.html#cb44-5"></a><span class="cf">for</span> e <span class="kw">in</span> SeqIO.parse(LIB<span class="op">/</span><span class="st">&quot;inputs/HMP2_MAG_00002-contigs.fa&quot;</span>, <span class="st">&quot;fasta&quot;</span>):</span>
<span id="cb44-6"><a href="module-3---lab-practical.html#cb44-6"></a>    MAG2.add(e.<span class="bu">id</span>)</span>
<span id="cb44-7"><a href="module-3---lab-practical.html#cb44-7"></a><span class="bu">len</span>(MAG1), <span class="bu">len</span>(MAG2)</span></code></pre></div>
<details>
<summary>
Python <code>set</code>
</summary>
<p>Sets enable easy checking of membership using <code>in</code>. For example:</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb45-1"><a href="module-3---lab-practical.html#cb45-1"></a>X <span class="op">=</span> {<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>}</span>
<span id="cb45-2"><a href="module-3---lab-practical.html#cb45-2"></a><span class="dv">1</span> <span class="kw">in</span> X <span class="co"># True</span></span>
<span id="cb45-3"><a href="module-3---lab-practical.html#cb45-3"></a><span class="dv">0</span> <span class="kw">in</span> X <span class="co"># False</span></span>
<span id="cb45-4"><a href="module-3---lab-practical.html#cb45-4"></a>empty_set <span class="op">=</span> <span class="bu">set</span>()</span></code></pre></div>
<hr />
</details>
<p>We can use the whitelists with additional parsing to get the family and MAG for each hit.</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb46-1"><a href="module-3---lab-practical.html#cb46-1"></a>_fams <span class="op">=</span> []</span>
<span id="cb46-2"><a href="module-3---lab-practical.html#cb46-2"></a>_mags <span class="op">=</span> []</span>
<span id="cb46-3"><a href="module-3---lab-practical.html#cb46-3"></a>df_fam <span class="op">=</span> pd.DataFrame(df_hits)</span>
<span id="cb46-4"><a href="module-3---lab-practical.html#cb46-4"></a><span class="cf">for</span> _, row <span class="kw">in</span> df_fam.iterrows():</span>
<span id="cb46-5"><a href="module-3---lab-practical.html#cb46-5"></a>    orf <span class="op">=</span> row[<span class="st">&quot;qseqid&quot;</span>]</span>
<span id="cb46-6"><a href="module-3---lab-practical.html#cb46-6"></a>    parts <span class="op">=</span> orf.split(<span class="st">&quot;_&quot;</span>)</span>
<span id="cb46-7"><a href="module-3---lab-practical.html#cb46-7"></a>    contig <span class="op">=</span> <span class="st">&quot;_&quot;</span>.join(parts[:<span class="op">-</span><span class="dv">1</span>])</span>
<span id="cb46-8"><a href="module-3---lab-practical.html#cb46-8"></a>    <span class="cf">if</span> contig <span class="kw">in</span> MAG1:</span>
<span id="cb46-9"><a href="module-3---lab-practical.html#cb46-9"></a>        mag <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb46-10"><a href="module-3---lab-practical.html#cb46-10"></a>    <span class="cf">elif</span> contig <span class="kw">in</span> MAG2:</span>
<span id="cb46-11"><a href="module-3---lab-practical.html#cb46-11"></a>        mag <span class="op">=</span> <span class="dv">2</span></span>
<span id="cb46-12"><a href="module-3---lab-practical.html#cb46-12"></a>    <span class="cf">else</span>:</span>
<span id="cb46-13"><a href="module-3---lab-practical.html#cb46-13"></a>        mag <span class="op">=</span> <span class="dv">0</span> <span class="co"># not in either MAG</span></span>
<span id="cb46-14"><a href="module-3---lab-practical.html#cb46-14"></a>    _mags.append(mag)</span>
<span id="cb46-15"><a href="module-3---lab-practical.html#cb46-15"></a></span>
<span id="cb46-16"><a href="module-3---lab-practical.html#cb46-16"></a>    description <span class="op">=</span> row[<span class="st">&quot;sseqid&quot;</span>]</span>
<span id="cb46-17"><a href="module-3---lab-practical.html#cb46-17"></a>    tcdb_id <span class="op">=</span> description.split(<span class="st">&quot;|&quot;</span>)[<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb46-18"><a href="module-3---lab-practical.html#cb46-18"></a>    parts <span class="op">=</span> tcdb_id.split(<span class="st">&quot;.&quot;</span>)</span>
<span id="cb46-19"><a href="module-3---lab-practical.html#cb46-19"></a>    family <span class="op">=</span> <span class="st">&quot;.&quot;</span>.join(parts[:<span class="dv">3</span>])</span>
<span id="cb46-20"><a href="module-3---lab-practical.html#cb46-20"></a>    _fams.append(family)</span>
<span id="cb46-21"><a href="module-3---lab-practical.html#cb46-21"></a></span>
<span id="cb46-22"><a href="module-3---lab-practical.html#cb46-22"></a>df_fam[<span class="st">&quot;family&quot;</span>] <span class="op">=</span> _fams</span>
<span id="cb46-23"><a href="module-3---lab-practical.html#cb46-23"></a>df_fam[<span class="st">&quot;mag&quot;</span>] <span class="op">=</span> _mags</span>
<span id="cb46-24"><a href="module-3---lab-practical.html#cb46-24"></a>df_fam <span class="op">=</span> df_fam[df_fam[<span class="st">&quot;mag&quot;</span>] <span class="op">!=</span> <span class="dv">0</span>] <span class="co"># remove no-MAG ORFs</span></span>
<span id="cb46-25"><a href="module-3---lab-practical.html#cb46-25"></a><span class="bu">print</span>(df_fam.shape)</span>
<span id="cb46-26"><a href="module-3---lab-practical.html#cb46-26"></a>df_fam.head(<span class="dv">2</span>)</span></code></pre></div>
<details>
<summary>
Python <code>dict</code>
</summary>
<p>Dictionaries enable the lookup of values by their corresponding key.</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb47-1"><a href="module-3---lab-practical.html#cb47-1"></a>d <span class="op">=</span> {<span class="st">&quot;a&quot;</span>: <span class="dv">1</span>, <span class="st">&quot;b&quot;</span>: <span class="dv">2</span>, <span class="st">&quot;c&quot;</span>: <span class="dv">3</span>}</span>
<span id="cb47-2"><a href="module-3---lab-practical.html#cb47-2"></a>d[<span class="st">&quot;a&quot;</span>] <span class="co"># 1</span></span></code></pre></div>
<p>By using <code>enumerate</code> to provide the index of each element in an iterable (set, list, dataframe, etc.),
dictionaries that map from values to their index can be created.</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb48-1"><a href="module-3---lab-practical.html#cb48-1"></a>values <span class="op">=</span> [<span class="st">&quot;a&quot;</span>, <span class="st">&quot;b&quot;</span>, <span class="st">&quot;c&quot;</span>]</span>
<span id="cb48-2"><a href="module-3---lab-practical.html#cb48-2"></a>value2i <span class="op">=</span> {value: i <span class="cf">for</span> i, value <span class="kw">in</span> <span class="bu">enumerate</span>(values)}</span>
<span id="cb48-3"><a href="module-3---lab-practical.html#cb48-3"></a>value2i[<span class="st">&quot;a&quot;</span>] <span class="co"># 0</span></span></code></pre></div>
<hr />
</details>
<p>To visualize the results, we will first organize the data into a matrix of counts for each transporter family and MAG.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb49-1"><a href="module-3---lab-practical.html#cb49-1"></a>mat <span class="op">=</span> np.zeros(shape<span class="op">=</span>(<span class="dv">2</span>, df_fam[<span class="st">&quot;family&quot;</span>].nunique())) <span class="co"># MAGs x families</span></span>
<span id="cb49-2"><a href="module-3---lab-practical.html#cb49-2"></a>mag2i <span class="op">=</span> {mag: i <span class="cf">for</span> i, mag <span class="kw">in</span> <span class="bu">enumerate</span>([<span class="dv">1</span>, <span class="dv">2</span>])}</span>
<span id="cb49-3"><a href="module-3---lab-practical.html#cb49-3"></a>fam2j <span class="op">=</span> {fam: i <span class="cf">for</span> i, fam <span class="kw">in</span> <span class="bu">enumerate</span>(df_fam[<span class="st">&quot;family&quot;</span>].unique())}</span>
<span id="cb49-4"><a href="module-3---lab-practical.html#cb49-4"></a><span class="cf">for</span> _, row <span class="kw">in</span> df_fam.groupby([<span class="st">&quot;family&quot;</span>, <span class="st">&quot;mag&quot;</span>])[[<span class="st">&quot;qseqid&quot;</span>]].count().reset_index().iterrows():</span>
<span id="cb49-5"><a href="module-3---lab-practical.html#cb49-5"></a>    i <span class="op">=</span> mag2i[row[<span class="st">&quot;mag&quot;</span>]]</span>
<span id="cb49-6"><a href="module-3---lab-practical.html#cb49-6"></a>    j <span class="op">=</span> fam2j[row[<span class="st">&quot;family&quot;</span>]]</span>
<span id="cb49-7"><a href="module-3---lab-practical.html#cb49-7"></a>    mat[i, j] <span class="op">=</span> row[<span class="st">&quot;qseqid&quot;</span>]</span>
<span id="cb49-8"><a href="module-3---lab-practical.html#cb49-8"></a>order <span class="op">=</span> np.argsort(<span class="op">-</span>mat.<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">0</span>)) <span class="co"># produces a sorted list of indexes</span></span></code></pre></div>
<p>Some utility functions are provided to make simple figures with <code>Ploty</code>.
Let’s use them to compare the abundance of transporter families in each MAG with a diverging bar plot.</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb50-1"><a href="module-3---lab-practical.html#cb50-1"></a><span class="im">from</span> local.figures.template <span class="im">import</span> BaseFigure, ApplyTemplate, go</span>
<span id="cb50-2"><a href="module-3---lab-practical.html#cb50-2"></a></span>
<span id="cb50-3"><a href="module-3---lab-practical.html#cb50-3"></a>family_labels <span class="op">=</span> np.array([tcdb_families[k] <span class="cf">for</span> k <span class="kw">in</span> fam2j.keys()])</span>
<span id="cb50-4"><a href="module-3---lab-practical.html#cb50-4"></a>TOP_K <span class="op">=</span> <span class="dv">10</span></span>
<span id="cb50-5"><a href="module-3---lab-practical.html#cb50-5"></a>fig <span class="op">=</span> BaseFigure()</span>
<span id="cb50-6"><a href="module-3---lab-practical.html#cb50-6"></a><span class="cf">for</span> mag <span class="kw">in</span> [<span class="dv">1</span>, <span class="dv">2</span>]:</span>
<span id="cb50-7"><a href="module-3---lab-practical.html#cb50-7"></a>    i <span class="op">=</span> mag2i[mag]</span>
<span id="cb50-8"><a href="module-3---lab-practical.html#cb50-8"></a>    sign <span class="op">=</span> <span class="op">-</span><span class="dv">1</span> <span class="cf">if</span> mag <span class="op">==</span> <span class="dv">1</span> <span class="cf">else</span> <span class="dv">1</span> <span class="co"># -1 for left bar</span></span>
<span id="cb50-9"><a href="module-3---lab-practical.html#cb50-9"></a>    fig.add_trace(</span>
<span id="cb50-10"><a href="module-3---lab-practical.html#cb50-10"></a>        go.Bar(</span>
<span id="cb50-11"><a href="module-3---lab-practical.html#cb50-11"></a>            x <span class="op">=</span> sign<span class="op">*</span>mat[i, order[:TOP_K]],     <span class="co"># length of each bar</span></span>
<span id="cb50-12"><a href="module-3---lab-practical.html#cb50-12"></a>            y <span class="op">=</span> family_labels[order[:TOP_K]],   <span class="co"># name of each row</span></span>
<span id="cb50-13"><a href="module-3---lab-practical.html#cb50-13"></a>            orientation<span class="op">=</span><span class="st">&#39;h&#39;</span>,</span>
<span id="cb50-14"><a href="module-3---lab-practical.html#cb50-14"></a>            name <span class="op">=</span> <span class="ss">f&quot;MAG</span><span class="sc">{</span>mag<span class="sc">}</span><span class="ss">&quot;</span>,</span>
<span id="cb50-15"><a href="module-3---lab-practical.html#cb50-15"></a>        )</span>
<span id="cb50-16"><a href="module-3---lab-practical.html#cb50-16"></a>    )</span>
<span id="cb50-17"><a href="module-3---lab-practical.html#cb50-17"></a>    <span class="co"># break # this will only draw the bars of one MAG. Uncomment &quot;break&quot; to see what each trace is drawing.</span></span>
<span id="cb50-18"><a href="module-3---lab-practical.html#cb50-18"></a></span>
<span id="cb50-19"><a href="module-3---lab-practical.html#cb50-19"></a>fig <span class="op">=</span> ApplyTemplate(</span>
<span id="cb50-20"><a href="module-3---lab-practical.html#cb50-20"></a>    fig,</span>
<span id="cb50-21"><a href="module-3---lab-practical.html#cb50-21"></a>    layout<span class="op">=</span><span class="bu">dict</span>(</span>
<span id="cb50-22"><a href="module-3---lab-practical.html#cb50-22"></a>        legend_orientation <span class="op">=</span><span class="st">&#39;h&#39;</span>,</span>
<span id="cb50-23"><a href="module-3---lab-practical.html#cb50-23"></a>        barmode<span class="op">=</span><span class="st">&quot;relative&quot;</span>,</span>
<span id="cb50-24"><a href="module-3---lab-practical.html#cb50-24"></a>        width<span class="op">=</span><span class="dv">1000</span>, height<span class="op">=</span><span class="dv">400</span>,</span>
<span id="cb50-25"><a href="module-3---lab-practical.html#cb50-25"></a>    ),</span>
<span id="cb50-26"><a href="module-3---lab-practical.html#cb50-26"></a>)</span>
<span id="cb50-27"><a href="module-3---lab-practical.html#cb50-27"></a>fig.show()</span></code></pre></div>
<details>
<summary>
How to change the figure to show the top 3 instead?
</summary>
<div class="sourceCode" id="cb51"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb51-1"><a href="module-3---lab-practical.html#cb51-1"></a>TOP_K <span class="op">=</span> <span class="dv">3</span></span></code></pre></div>
<hr />
</details>
<details>
<summary>
Can you put MAG1 on the right and MAG2 on the left?
</summary>
<div class="sourceCode" id="cb52"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb52-1"><a href="module-3---lab-practical.html#cb52-1"></a>    sign <span class="op">=</span> <span class="op">-</span><span class="dv">1</span> <span class="cf">if</span> mag <span class="op">==</span> <span class="dv">2</span> <span class="cf">else</span> <span class="dv">1</span></span></code></pre></div>
<hr />
</details>
</div>
<div id="bakta" class="section level3 hasAnchor">
<h3>Bakta<a href="module-3---lab-practical.html#bakta" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><a href="https://quay.io/repository/biocontainers/bakta?tab=tags" class="uri">https://quay.io/repository/biocontainers/bakta?tab=tags</a></p>
<div class="sourceCode" id="cb53"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb53-1"><a href="module-3---lab-practical.html#cb53-1"></a><span class="co"># docker://quay.io/biocontainers/bakta:1.9.4--pyhdfd78af_0</span></span>
<span id="cb53-2"><a href="module-3---lab-practical.html#cb53-2"></a>bakta_sif <span class="op">=</span> LIB<span class="op">/</span><span class="st">&quot;bakta.sif&quot;</span></span></code></pre></div>
<p>Bakta is an automated workflow for annotating both coding and non-coding features using a variety of tools [1].
Its core protocol can be reduced to ORF prediction by Prodigal, followed by sequence similary search by Diamond, as we have just done.
Many additional tools and consolidated reference databases are used to search for non-coding features, provide information on ORFs with no good hits in any database, and perform QC.</p>
<ol style="list-style-type: decimal">
<li>Schwengers O, Jelonek L, Dieckmann MA, Beyvers S, Blom J, Goesmann A. Bakta: rapid and standardized annotation of bacterial genomes via alignment-free sequence identification. Microb Genomics. 2021;7(11):000685. <a href="https://doi.org/10.1099/mgen.0.000685" class="uri">https://doi.org/10.1099/mgen.0.000685</a></li>
</ol>
<p>To run Bakta, we expect to first set up reference databases before it will:
- accept contigs, for which it will first run Prodigal for us
- or accept a list of regions + contigs, for which Prodigal will be skipped
- and produce many files containing useful annotations</p>
<details>
<summary>
Why is a protein sequence fasta insufficient to full utilize Bakta?
</summary>
<p>Non-coding features would not be represented.</p>
<hr />
</details>
<p>The database was pre-installed, but here is some code to guide you if you decide to revisit in the future.</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb54-1"><a href="module-3---lab-practical.html#cb54-1"></a>Shell(<span class="ss">f&quot;apptainer exec </span><span class="sc">{</span>bakta_sif<span class="sc">}</span><span class="ss"> bakta_db list&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb55"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb55-1"><a href="module-3---lab-practical.html#cb55-1"></a><span class="co"># bakta_db = LIB/&quot;bakta_db/db&quot;        # 2 hours, ~70 GB</span></span>
<span id="cb55-2"><a href="module-3---lab-practical.html#cb55-2"></a>bakta_db <span class="op">=</span> LIB<span class="op">/</span><span class="st">&quot;bakta_db/db-light&quot;</span>  <span class="co"># 30 mins, ~3.5 GB</span></span>
<span id="cb55-3"><a href="module-3---lab-practical.html#cb55-3"></a><span class="cf">if</span> <span class="kw">not</span> bakta_db.exists():</span>
<span id="cb55-4"><a href="module-3---lab-practical.html#cb55-4"></a>    <span class="cf">if</span> <span class="st">&quot;light&quot;</span> <span class="kw">in</span> <span class="bu">str</span>(bakta_db):</span>
<span id="cb55-5"><a href="module-3---lab-practical.html#cb55-5"></a>        _type <span class="op">=</span> <span class="st">&quot;light&quot;</span></span>
<span id="cb55-6"><a href="module-3---lab-practical.html#cb55-6"></a>    <span class="cf">else</span>:</span>
<span id="cb55-7"><a href="module-3---lab-practical.html#cb55-7"></a>        _type <span class="op">=</span> <span class="st">&quot;full&quot;</span></span>
<span id="cb55-8"><a href="module-3---lab-practical.html#cb55-8"></a>    Shell(<span class="ss">f&quot;apptainer exec </span><span class="sc">{</span>bakta_sif<span class="sc">}</span><span class="ss"> bakta_db download --output </span><span class="sc">{</span>bakta_db<span class="sc">}</span><span class="ss"> --type </span><span class="sc">{</span>_type<span class="sc">}</span><span class="ss">&quot;</span>)</span></code></pre></div>
<p>There may be other annotation tools that we wish to run in addition to Bakta, which depend on predicted ORFs.
Waiting for Bakta to finish before we recieve the predicted ORFs may be inconvenient, especially for large metagenomic datasets.
Running Prodigal ourselves and passing the ORFs to Bakta frees us to run additional tools in parallel. Bakta does not accept ORFs
that run off the edge of contigs, so we will need to filter these out.</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb56-1"><a href="module-3---lab-practical.html#cb56-1"></a>gff <span class="op">=</span> pd.read_csv(prodigal_gff, sep<span class="op">=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, header<span class="op">=</span><span class="va">None</span>, comment<span class="op">=</span><span class="st">&quot;#&quot;</span>)</span>
<span id="cb56-2"><a href="module-3---lab-practical.html#cb56-2"></a>complete_orfs <span class="op">=</span> [] <span class="co"># a boolean mask</span></span>
<span id="cb56-3"><a href="module-3---lab-practical.html#cb56-3"></a><span class="cf">for</span> _, row <span class="kw">in</span> gff.iterrows():</span>
<span id="cb56-4"><a href="module-3---lab-practical.html#cb56-4"></a>    meta <span class="op">=</span> <span class="bu">tuple</span>(row)[<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb56-5"><a href="module-3---lab-practical.html#cb56-5"></a>    <span class="cf">if</span> <span class="st">&quot;partial=00&quot;</span> <span class="kw">in</span> meta:</span>
<span id="cb56-6"><a href="module-3---lab-practical.html#cb56-6"></a>        complete_orfs.append(<span class="va">True</span>)</span>
<span id="cb56-7"><a href="module-3---lab-practical.html#cb56-7"></a>    <span class="cf">else</span>:</span>
<span id="cb56-8"><a href="module-3---lab-practical.html#cb56-8"></a>        complete_orfs.append(<span class="va">False</span>)</span>
<span id="cb56-9"><a href="module-3---lab-practical.html#cb56-9"></a></span>
<span id="cb56-10"><a href="module-3---lab-practical.html#cb56-10"></a>gff_complete <span class="op">=</span> <span class="st">&quot;./prodigal/orfs.complete.gff&quot;</span></span>
<span id="cb56-11"><a href="module-3---lab-practical.html#cb56-11"></a>gff[complete_orfs].to_csv(gff_complete, sep<span class="op">=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, header<span class="op">=</span><span class="va">False</span>, index<span class="op">=</span><span class="va">False</span>)</span></code></pre></div>
<p>Running Bakta, even on this reduced dataset, may take <strong>15 minutes or more</strong>. Precomputed outputs are available at <code>{LIB}/outputs/bakta</code>.</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb57-1"><a href="module-3---lab-practical.html#cb57-1"></a>Shell(<span class="ss">f&quot;&quot;&quot;</span></span>
<span id="cb57-2"><a href="module-3---lab-practical.html#cb57-2"></a><span class="ss">apptainer exec </span><span class="sc">{</span>bakta_sif<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb57-3"><a href="module-3---lab-practical.html#cb57-3"></a><span class="ss">    bakta --meta --threads </span><span class="sc">{</span>CPUS<span class="sc">}</span><span class="ss"> --force --skip-plot </span><span class="op">\</span></span>
<span id="cb57-4"><a href="module-3---lab-practical.html#cb57-4"></a><span class="ss">        --db </span><span class="sc">{</span>bakta_db<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb57-5"><a href="module-3---lab-practical.html#cb57-5"></a><span class="ss">        --keep-contig-headers </span><span class="op">\</span></span>
<span id="cb57-6"><a href="module-3---lab-practical.html#cb57-6"></a><span class="ss">        --output ./bakta/ </span><span class="op">\</span></span>
<span id="cb57-7"><a href="module-3---lab-practical.html#cb57-7"></a><span class="ss">        --regions </span><span class="sc">{</span>gff_complete<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb57-8"><a href="module-3---lab-practical.html#cb57-8"></a><span class="ss">        </span><span class="sc">{</span>LIB<span class="sc">}</span><span class="ss">/inputs/assembly.fa</span></span>
<span id="cb57-9"><a href="module-3---lab-practical.html#cb57-9"></a><span class="ss">&quot;&quot;&quot;</span>)</span></code></pre></div>
</div>
<div id="resistance-gene-identifier-rgi" class="section level3 hasAnchor">
<h3>Resistance Gene Identifier (RGI)<a href="module-3---lab-practical.html#resistance-gene-identifier-rgi" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><a href="https://quay.io/repository/biocontainers/rgi?tab=tags" class="uri">https://quay.io/repository/biocontainers/rgi?tab=tags</a></p>
<div class="sourceCode" id="cb58"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb58-1"><a href="module-3---lab-practical.html#cb58-1"></a><span class="co"># docker://quay.io/biocontainers/rgi:6.0.4--pyh05cac1d_0</span></span>
<span id="cb58-2"><a href="module-3---lab-practical.html#cb58-2"></a>rgi_sif <span class="op">=</span> LIB<span class="op">/</span><span class="st">&quot;rgi.sif&quot;</span></span></code></pre></div>
<p>Sometimes, specific functions are of interest, such as antibiotic resistance.
RGI is a tool for searching the Comprehensive Antibiotic Resistance Database (CARD) [1, 2] using tuned protocols for Diamond.</p>
<ol style="list-style-type: decimal">
<li>McArthur AG, Waglechner N, Nizam F, Yan A, Azad MA, Baylay AJ, et al. The Comprehensive Antibiotic Resistance Database. Antimicrob Agents Chemother. 2013;57(7):3348–57. <a href="https://doi.org/10.1128/aac.00419-13" class="uri">https://doi.org/10.1128/aac.00419-13</a></li>
<li>Alcock BP, Huynh W, Chalil R, Smith KW, Raphenya AR, Wlodarski MA, et al. CARD 2023: expanded curation, support for machine learning, and resistome prediction at the Comprehensive Antibiotic Resistance Database. Nucleic Acids Res. 2023;51(D1):D690–9. <a href="https://doi.org/10.1093/nar/gkac920" class="uri">https://doi.org/10.1093/nar/gkac920</a></li>
</ol>
<div class="sourceCode" id="cb59"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb59-1"><a href="module-3---lab-practical.html#cb59-1"></a>Shell(<span class="ss">f&quot;apptainer exec </span><span class="sc">{</span>rgi_sif<span class="sc">}</span><span class="ss"> rgi --help&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb60"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb60-1"><a href="module-3---lab-practical.html#cb60-1"></a>Shell(<span class="ss">f&quot;apptainer exec </span><span class="sc">{</span>rgi_sif<span class="sc">}</span><span class="ss"> rgi main --help&quot;</span>)</span></code></pre></div>
<p>Errors are a common occurrence for bioinformatics tools. The following is a command which appears to be correct,
but will nonetheless fail. Read the error messages and use the hints to fix the issue. Additional warnings will appear
that do not crash the program. These can be ignored. RGI also provides no indication of progress or successful completion,
so checking the output folder will be necessary.</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb61-1"><a href="module-3---lab-practical.html#cb61-1"></a>rgi_out <span class="op">=</span> Path(<span class="st">&quot;./rgi/chicken&quot;</span>)</span>
<span id="cb61-2"><a href="module-3---lab-practical.html#cb61-2"></a>Shell(<span class="ss">f&quot;&quot;&quot;</span></span>
<span id="cb61-3"><a href="module-3---lab-practical.html#cb61-3"></a><span class="ss">apptainer exec </span><span class="sc">{</span>rgi_sif<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb61-4"><a href="module-3---lab-practical.html#cb61-4"></a><span class="ss">    rgi main </span><span class="op">\</span></span>
<span id="cb61-5"><a href="module-3---lab-practical.html#cb61-5"></a><span class="ss">        --num_threads </span><span class="sc">{</span>CPUS<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb61-6"><a href="module-3---lab-practical.html#cb61-6"></a><span class="ss">        --clean </span><span class="op">\</span></span>
<span id="cb61-7"><a href="module-3---lab-practical.html#cb61-7"></a><span class="ss">        -i </span><span class="sc">{</span>prodigal_aa<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb61-8"><a href="module-3---lab-practical.html#cb61-8"></a><span class="ss">        -o </span><span class="sc">{</span>rgi_out<span class="sc">}</span></span>
<span id="cb61-9"><a href="module-3---lab-practical.html#cb61-9"></a><span class="ss">&quot;&quot;&quot;</span>)</span></code></pre></div>
<details>
<summary>
<code>FileNotFoundError</code> hint
</summary>
Is the path to the missing file related to any of the paths we gave it?
</details>
<details>
<summary>
<code>FileNotFoundError</code> explanation
</summary>
We indicated an output path within a new folder <code>rgi_out = Path("./rgi/chicken")</code>. It appears that RGI can’t make this folder for us,
so we will have to create it ourselves with <code>mkdir -p {rgi_out.parent}</code>. <code>parent</code> refers to the parent folder.
</details>
<details>
<summary>
<code>invalid fasta</code> hint
</summary>
Nucleotide?
</details>
<details>
<summary>
<code>invalid fasta</code> explanation
</summary>
We can use the argument <code>-t protein</code> to indicate that the input is a protein sequence.
</details>
<details>
<summary>
solution
</summary>
<div class="sourceCode" id="cb62"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb62-1"><a href="module-3---lab-practical.html#cb62-1"></a>rgi_out <span class="op">=</span> Path(<span class="st">&quot;./rgi/chicken&quot;</span>)</span>
<span id="cb62-2"><a href="module-3---lab-practical.html#cb62-2"></a>Shell(<span class="ss">f&quot;&quot;&quot;</span></span>
<span id="cb62-3"><a href="module-3---lab-practical.html#cb62-3"></a><span class="ss">mkdir -p </span><span class="sc">{</span>rgi_out<span class="sc">.</span>parent<span class="sc">}</span></span>
<span id="cb62-4"><a href="module-3---lab-practical.html#cb62-4"></a><span class="ss">apptainer exec </span><span class="sc">{</span>rgi_sif<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb62-5"><a href="module-3---lab-practical.html#cb62-5"></a><span class="ss">    rgi main </span><span class="op">\</span></span>
<span id="cb62-6"><a href="module-3---lab-practical.html#cb62-6"></a><span class="ss">        --num_threads </span><span class="sc">{</span>CPUS<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb62-7"><a href="module-3---lab-practical.html#cb62-7"></a><span class="ss">        --clean </span><span class="op">\</span></span>
<span id="cb62-8"><a href="module-3---lab-practical.html#cb62-8"></a><span class="ss">        -t protein </span><span class="op">\</span></span>
<span id="cb62-9"><a href="module-3---lab-practical.html#cb62-9"></a><span class="ss">        -i </span><span class="sc">{</span>prodigal_aa<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb62-10"><a href="module-3---lab-practical.html#cb62-10"></a><span class="ss">        -o </span><span class="sc">{</span>rgi_out<span class="sc">}</span></span>
<span id="cb62-11"><a href="module-3---lab-practical.html#cb62-11"></a><span class="ss">&quot;&quot;&quot;</span>)</span>
<span id="cb62-12"><a href="module-3---lab-practical.html#cb62-12"></a>rgi_out <span class="op">=</span> rgi_out.with_suffix(<span class="st">&quot;.txt&quot;</span>) <span class="co"># RGI adds the .txt suffix</span></span>
<span id="cb62-13"><a href="module-3---lab-practical.html#cb62-13"></a><span class="co"># or</span></span>
<span id="cb62-14"><a href="module-3---lab-practical.html#cb62-14"></a><span class="co"># rgi_out = Path(&quot;./rgi/chicken.txt&quot;)</span></span></code></pre></div>
<hr />
</details>
<p>Let’s take a quick peak at the results.</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb63-1"><a href="module-3---lab-practical.html#cb63-1"></a>Shell(<span class="ss">f&quot;&quot;&quot;</span></span>
<span id="cb63-2"><a href="module-3---lab-practical.html#cb63-2"></a><span class="ss">_df = pd.read_csv(rgi_out, sep=&quot;</span><span class="ch">\t</span><span class="ss">&quot;, header=0)</span></span>
<span id="cb63-3"><a href="module-3---lab-practical.html#cb63-3"></a><span class="ss">_df</span></span>
<span id="cb63-4"><a href="module-3---lab-practical.html#cb63-4"></a><span class="ss">&quot;&quot;&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="machine-learning" class="section level2 hasAnchor">
<h2>Machine learning<a href="module-3---lab-practical.html#machine-learning" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Alignment-based approaches that assign functional annotations based on sequence similarity rely on databases of known sequnces
to transfer their known functions to the ORFs of a given sample. Machine learning may be able to discover more general patterns
that can extend annotations to previously unseen sequences.</p>
<div id="kofamscan" class="section level3 hasAnchor">
<h3>Kofamscan<a href="module-3---lab-practical.html#kofamscan" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><a href="https://quay.io/repository/hallamlab/external_kofamscan?tab=tags" class="uri">https://quay.io/repository/hallamlab/external_kofamscan?tab=tags</a></p>
<div class="sourceCode" id="cb64"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb64-1"><a href="module-3---lab-practical.html#cb64-1"></a><span class="co"># docker://quay.io/hallamlab/external_kofamscan:1.3.0</span></span>
<span id="cb64-2"><a href="module-3---lab-practical.html#cb64-2"></a>kofamscan_sif <span class="op">=</span> LIB<span class="op">/</span><span class="st">&quot;kofamscan.sif&quot;</span></span></code></pre></div>
<p>Kofamscan uses a machine learning technique called hidden markov models to learn the amino acid sequence patterns
accociated with various functions, organized in the KEGG database [1, 2]. Kofamscan sits somewhere in the middle
between alignment-based and machine learning-based annotation methods.</p>
<ol style="list-style-type: decimal">
<li>Aramaki T, Blanc-Mathieu R, Endo H, Ohkubo K, Kanehisa M, Goto S, et al. KofamKOALA: KEGG Ortholog assignment based on profile HMM and adaptive score threshold. Bioinformatics. 2020;36(7):2251–2. <a href="https://doi.org/10.1093/bioinformatics/btz859" class="uri">https://doi.org/10.1093/bioinformatics/btz859</a></li>
<li>Kanehisa M, Goto S. KEGG: Kyoto Encyclopedia of Genes and Genomes. Nucleic Acids Res. 2000;28(1):27–30. <a href="https://doi.org/10.1093/nar/28.1.27" class="uri">https://doi.org/10.1093/nar/28.1.27</a></li>
</ol>
<div class="sourceCode" id="cb65"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb65-1"><a href="module-3---lab-practical.html#cb65-1"></a>Shell(<span class="ss">f&quot;apptainer exec </span><span class="sc">{</span>kofamscan_sif<span class="sc">}</span><span class="ss"> kofamscan --help&quot;</span>) <span class="co"># expect crash</span></span></code></pre></div>
<p>It looks like we can’t even get a help message with a naiive call to the tool.
Let’s check the documentation for some clues. Specifically, there is short section marked <code>## Usage</code> that will be helpful.</p>
<p><a href="https://www.genome.jp/ftp/tools/kofam_scan/README.md" class="uri">https://www.genome.jp/ftp/tools/kofam_scan/README.md</a></p>
<p>Ah, so the executable is <code>exec_annotation</code>.</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb66-1"><a href="module-3---lab-practical.html#cb66-1"></a>Shell(<span class="ss">f&quot;apptainer exec </span><span class="sc">{</span>kofamscan_sif<span class="sc">}</span><span class="ss"> exec_annotation --help&quot;</span>)</span></code></pre></div>
<p>Kofamscan needs pre-trained HMM models, which have been downloaded from these links.</p>
<ul>
<li><code>ftp://ftp.genome.jp/pub/db/kofam/ko_list.gz</code></li>
<li><code>ftp://ftp.genome.jp/pub/db/kofam/profiles.tar.gz</code></li>
</ul>
<p>Kofamscan will take a long time to run. You may want to try running it on the hypothetical ORFs from the bakta output at
<code>LIB/"outputs/bakta.full/assembly.hypotheticals.faa</code> instead. Pre-computed outputs are available at <code>LIB/"outputs/kofamscan/kos.out"</code></p>
<div class="sourceCode" id="cb67"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb67-1"><a href="module-3---lab-practical.html#cb67-1"></a>kofamscan_raw <span class="op">=</span> Path(<span class="st">&quot;./kofamscan/kos.out&quot;</span>)</span>
<span id="cb67-2"><a href="module-3---lab-practical.html#cb67-2"></a>Shell(<span class="ss">f&quot;&quot;&quot;</span></span>
<span id="cb67-3"><a href="module-3---lab-practical.html#cb67-3"></a><span class="ss">mkdir -p ./kofamscan</span></span>
<span id="cb67-4"><a href="module-3---lab-practical.html#cb67-4"></a><span class="ss">apptainer exec </span><span class="sc">{</span>kofamscan_sif<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb67-5"><a href="module-3---lab-practical.html#cb67-5"></a><span class="ss">    exec_annotation </span><span class="op">\</span></span>
<span id="cb67-6"><a href="module-3---lab-practical.html#cb67-6"></a><span class="ss">        --cpu=</span><span class="sc">{</span>CPUS<span class="sc">}</span><span class="ss"> --format=detail --no-report-unannotated </span><span class="op">\</span></span>
<span id="cb67-7"><a href="module-3---lab-practical.html#cb67-7"></a><span class="ss">        --profile=</span><span class="sc">{</span>LIB<span class="sc">}</span><span class="ss">/kofamscan_db/profiles/prokaryote.hal </span><span class="op">\</span></span>
<span id="cb67-8"><a href="module-3---lab-practical.html#cb67-8"></a><span class="ss">        --ko-list=</span><span class="sc">{</span>LIB<span class="sc">}</span><span class="ss">/kofamscan_db/ko_list </span><span class="op">\</span></span>
<span id="cb67-9"><a href="module-3---lab-practical.html#cb67-9"></a><span class="ss">        -o </span><span class="sc">{</span>kofamscan_raw<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb67-10"><a href="module-3---lab-practical.html#cb67-10"></a><span class="ss">        </span><span class="sc">{</span>prodigal_aa<span class="sc">}</span></span>
<span id="cb67-11"><a href="module-3---lab-practical.html#cb67-11"></a><span class="ss">&quot;&quot;&quot;</span>)</span></code></pre></div>
<p>Kofamscan attempts to assigns KEGG orthology (KO) numbers to each ORF, linking them to rich KEGG databases
which provide hierachical organization and metabolic context. We can use the provided utility to parse the
results into a partial model of KEGG for local use.</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb68-1"><a href="module-3---lab-practical.html#cb68-1"></a><span class="im">from</span> local.models.kegg_orthology <span class="im">import</span> ParseKofamScanResults</span>
<span id="cb68-2"><a href="module-3---lab-practical.html#cb68-2"></a></span>
<span id="cb68-3"><a href="module-3---lab-practical.html#cb68-3"></a>model <span class="op">=</span> ParseKofamScanResults(</span>
<span id="cb68-4"><a href="module-3---lab-practical.html#cb68-4"></a>    kofamscan_raw,</span>
<span id="cb68-5"><a href="module-3---lab-practical.html#cb68-5"></a>    LIB<span class="op">/</span><span class="st">&quot;kofamscan_db/api_kegg.db&quot;</span>,</span>
<span id="cb68-6"><a href="module-3---lab-practical.html#cb68-6"></a>    LIB<span class="op">/</span><span class="st">&quot;kofamscan_db/brite.json&quot;</span>,</span>
<span id="cb68-7"><a href="module-3---lab-practical.html#cb68-7"></a>)</span></code></pre></div>
<div class="sourceCode" id="cb69"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb69-1"><a href="module-3---lab-practical.html#cb69-1"></a>model.Summary()</span></code></pre></div>
<div class="sourceCode" id="cb70"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb70-1"><a href="module-3---lab-practical.html#cb70-1"></a>df <span class="op">=</span> model.df</span>
<span id="cb70-2"><a href="module-3---lab-practical.html#cb70-2"></a>df <span class="op">=</span> df[df.score <span class="op">&gt;</span> df.hmm_threshold] <span class="co"># retrieve only assignments that pass the model&#39;s adaptive confidence threshold</span></span>
<span id="cb70-3"><a href="module-3---lab-practical.html#cb70-3"></a><span class="bu">print</span>(df.shape)</span>
<span id="cb70-4"><a href="module-3---lab-practical.html#cb70-4"></a>df.head(<span class="dv">2</span>)</span></code></pre></div>
<p>Let’s take a peek at a KEGG database entry.</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb71-1"><a href="module-3---lab-practical.html#cb71-1"></a>model.GetComponentModel(<span class="st">&quot;K06442&quot;</span>)._raw</span></code></pre></div>
<p>We can also look at where the assigned function sits in the BRITE functional hierarchy.</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb72-1"><a href="module-3---lab-practical.html#cb72-1"></a><span class="cf">for</span> lineage <span class="kw">in</span> model.GetLineage(<span class="st">&quot;K06442&quot;</span>):</span>
<span id="cb72-2"><a href="module-3---lab-practical.html#cb72-2"></a>    <span class="cf">for</span> k <span class="kw">in</span> lineage:</span>
<span id="cb72-3"><a href="module-3---lab-practical.html#cb72-3"></a>        <span class="bu">print</span>(model.GetNodeInfo(k))</span></code></pre></div>
<p><code>GetCategories</code> provides a shorthand to retrieve 2 useful levels from the BRITE heriarchy.</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb73-1"><a href="module-3---lab-practical.html#cb73-1"></a><span class="cf">for</span> l2, l1 <span class="kw">in</span> model.GetCategories(<span class="st">&quot;K06442&quot;</span>):</span>
<span id="cb73-2"><a href="module-3---lab-practical.html#cb73-2"></a>    <span class="cf">for</span> k <span class="kw">in</span> [l2, l1]:</span>
<span id="cb73-3"><a href="module-3---lab-practical.html#cb73-3"></a>        <span class="bu">print</span>(model.GetNodeInfo(k))</span></code></pre></div>
<p>Just like what we did with transporters, let’s visualize the abundance of functions aggregated by BRITE
categories for each MAG. First, we will compile a count of each category by MAG.</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb74-1"><a href="module-3---lab-practical.html#cb74-1"></a>rows <span class="op">=</span> []</span>
<span id="cb74-2"><a href="module-3---lab-practical.html#cb74-2"></a><span class="cf">for</span> _, row <span class="kw">in</span> df.iterrows():</span>
<span id="cb74-3"><a href="module-3---lab-practical.html#cb74-3"></a>    ko <span class="op">=</span> row[<span class="st">&quot;ko&quot;</span>]</span>
<span id="cb74-4"><a href="module-3---lab-practical.html#cb74-4"></a>    orf <span class="op">=</span> row[<span class="st">&quot;orf&quot;</span>]</span>
<span id="cb74-5"><a href="module-3---lab-practical.html#cb74-5"></a>    contig <span class="op">=</span> <span class="st">&quot;_&quot;</span>.join(orf.split(<span class="st">&quot;_&quot;</span>)[:<span class="op">-</span><span class="dv">1</span>])</span>
<span id="cb74-6"><a href="module-3---lab-practical.html#cb74-6"></a>    <span class="cf">if</span> contig <span class="kw">in</span> MAG1:</span>
<span id="cb74-7"><a href="module-3---lab-practical.html#cb74-7"></a>        mag <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb74-8"><a href="module-3---lab-practical.html#cb74-8"></a>    <span class="cf">elif</span> contig <span class="kw">in</span> MAG2:</span>
<span id="cb74-9"><a href="module-3---lab-practical.html#cb74-9"></a>        mag <span class="op">=</span> <span class="dv">2</span></span>
<span id="cb74-10"><a href="module-3---lab-practical.html#cb74-10"></a>    <span class="cf">else</span>:</span>
<span id="cb74-11"><a href="module-3---lab-practical.html#cb74-11"></a>        mag <span class="op">=</span> <span class="dv">0</span> <span class="co"># not in either MAG</span></span>
<span id="cb74-12"><a href="module-3---lab-practical.html#cb74-12"></a>    <span class="cf">for</span> category, _ <span class="kw">in</span> model.GetCategories(ko):</span>
<span id="cb74-13"><a href="module-3---lab-practical.html#cb74-13"></a>        rows.append((mag, category))</span>
<span id="cb74-14"><a href="module-3---lab-practical.html#cb74-14"></a>df_temp <span class="op">=</span> pd.DataFrame(rows, columns<span class="op">=</span>[<span class="st">&quot;mag&quot;</span>, <span class="st">&quot;category&quot;</span>])</span></code></pre></div>
<p>Then, we can format it as a matrix for plotting.</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb75-1"><a href="module-3---lab-practical.html#cb75-1"></a><span class="co"># create some index lookup dictionaries for the matrix</span></span>
<span id="cb75-2"><a href="module-3---lab-practical.html#cb75-2"></a>c2i <span class="op">=</span> {c: i <span class="cf">for</span> i, c <span class="kw">in</span> <span class="bu">enumerate</span>(df_temp[<span class="st">&quot;category&quot;</span>].unique())}    <span class="co"># category to index</span></span>
<span id="cb75-3"><a href="module-3---lab-practical.html#cb75-3"></a>mag2i <span class="op">=</span> {mag: i <span class="cf">for</span> i, mag <span class="kw">in</span> <span class="bu">enumerate</span>([<span class="dv">1</span>, <span class="dv">2</span>])}                    <span class="co"># mag to index</span></span>
<span id="cb75-4"><a href="module-3---lab-practical.html#cb75-4"></a>mat <span class="op">=</span> np.zeros(shape<span class="op">=</span>(<span class="bu">len</span>(mag2i), <span class="bu">len</span>(c2i))) <span class="co"># MAG x categories</span></span>
<span id="cb75-5"><a href="module-3---lab-practical.html#cb75-5"></a><span class="co"># iterate through the counts for each mag</span></span>
<span id="cb75-6"><a href="module-3---lab-practical.html#cb75-6"></a><span class="co"># assigning them to the correct cell using the index lookup dictionaries</span></span>
<span id="cb75-7"><a href="module-3---lab-practical.html#cb75-7"></a><span class="cf">for</span> _, row <span class="kw">in</span> df_temp.iterrows():</span>
<span id="cb75-8"><a href="module-3---lab-practical.html#cb75-8"></a>    mag <span class="op">=</span> row[<span class="st">&quot;mag&quot;</span>]</span>
<span id="cb75-9"><a href="module-3---lab-practical.html#cb75-9"></a>    <span class="cf">if</span> mag <span class="op">==</span> <span class="dv">0</span>: <span class="cf">continue</span></span>
<span id="cb75-10"><a href="module-3---lab-practical.html#cb75-10"></a>    i <span class="op">=</span> mag2i[mag]</span>
<span id="cb75-11"><a href="module-3---lab-practical.html#cb75-11"></a>    j <span class="op">=</span> c2i[row[<span class="st">&quot;category&quot;</span>]]</span>
<span id="cb75-12"><a href="module-3---lab-practical.html#cb75-12"></a>    mat[i, j] <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb75-13"><a href="module-3---lab-practical.html#cb75-13"></a></span>
<span id="cb75-14"><a href="module-3---lab-practical.html#cb75-14"></a>order <span class="op">=</span> np.argsort(<span class="op">-</span>mat.<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">0</span>)) <span class="co"># get the order of categories by size (largest -&gt; smallest)</span></span>
<span id="cb75-15"><a href="module-3---lab-practical.html#cb75-15"></a>mat.<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>)</span></code></pre></div>
<p>Plot the top 10 largest categories.</p>
<div class="sourceCode" id="cb76"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb76-1"><a href="module-3---lab-practical.html#cb76-1"></a><span class="im">from</span> local.figures.template <span class="im">import</span> BaseFigure, ApplyTemplate, go</span>
<span id="cb76-2"><a href="module-3---lab-practical.html#cb76-2"></a></span>
<span id="cb76-3"><a href="module-3---lab-practical.html#cb76-3"></a>category_labels <span class="op">=</span> np.array([model.GetNodeInfo(k)[<span class="dv">1</span>] <span class="cf">for</span> k <span class="kw">in</span> c2i.keys()])</span>
<span id="cb76-4"><a href="module-3---lab-practical.html#cb76-4"></a>K <span class="op">=</span> <span class="dv">10</span></span>
<span id="cb76-5"><a href="module-3---lab-practical.html#cb76-5"></a>fig <span class="op">=</span> BaseFigure()</span>
<span id="cb76-6"><a href="module-3---lab-practical.html#cb76-6"></a><span class="cf">for</span> mag <span class="kw">in</span> [<span class="dv">1</span>, <span class="dv">2</span>]:</span>
<span id="cb76-7"><a href="module-3---lab-practical.html#cb76-7"></a>    i <span class="op">=</span> mag2i[mag]</span>
<span id="cb76-8"><a href="module-3---lab-practical.html#cb76-8"></a>    sign <span class="op">=</span> <span class="op">-</span><span class="dv">1</span> <span class="cf">if</span> mag <span class="op">==</span> <span class="dv">1</span> <span class="cf">else</span> <span class="dv">1</span></span>
<span id="cb76-9"><a href="module-3---lab-practical.html#cb76-9"></a>    fig.add_trace(</span>
<span id="cb76-10"><a href="module-3---lab-practical.html#cb76-10"></a>        go.Bar(</span>
<span id="cb76-11"><a href="module-3---lab-practical.html#cb76-11"></a>            x <span class="op">=</span> sign<span class="op">*</span>mat[i, order[:K]],</span>
<span id="cb76-12"><a href="module-3---lab-practical.html#cb76-12"></a>            y <span class="op">=</span> category_labels[order[:K]],</span>
<span id="cb76-13"><a href="module-3---lab-practical.html#cb76-13"></a>            orientation<span class="op">=</span><span class="st">&#39;h&#39;</span>,</span>
<span id="cb76-14"><a href="module-3---lab-practical.html#cb76-14"></a>            name <span class="op">=</span> <span class="ss">f&quot;MAG</span><span class="sc">{</span>mag<span class="sc">}</span><span class="ss">&quot;</span>,</span>
<span id="cb76-15"><a href="module-3---lab-practical.html#cb76-15"></a>        )</span>
<span id="cb76-16"><a href="module-3---lab-practical.html#cb76-16"></a>    )</span>
<span id="cb76-17"><a href="module-3---lab-practical.html#cb76-17"></a></span>
<span id="cb76-18"><a href="module-3---lab-practical.html#cb76-18"></a>fig <span class="op">=</span> ApplyTemplate(</span>
<span id="cb76-19"><a href="module-3---lab-practical.html#cb76-19"></a>    fig,</span>
<span id="cb76-20"><a href="module-3---lab-practical.html#cb76-20"></a>    layout<span class="op">=</span><span class="bu">dict</span>(</span>
<span id="cb76-21"><a href="module-3---lab-practical.html#cb76-21"></a>        legend_orientation <span class="op">=</span><span class="st">&#39;h&#39;</span>,</span>
<span id="cb76-22"><a href="module-3---lab-practical.html#cb76-22"></a>        barmode<span class="op">=</span><span class="st">&quot;relative&quot;</span>,</span>
<span id="cb76-23"><a href="module-3---lab-practical.html#cb76-23"></a>        width<span class="op">=</span><span class="dv">1000</span>, height<span class="op">=</span><span class="dv">400</span>,</span>
<span id="cb76-24"><a href="module-3---lab-practical.html#cb76-24"></a>    ),</span>
<span id="cb76-25"><a href="module-3---lab-practical.html#cb76-25"></a>)</span>
<span id="cb76-26"><a href="module-3---lab-practical.html#cb76-26"></a>fig.show()</span></code></pre></div>
</div>
<div id="deepec" class="section level3 hasAnchor">
<h3>deepEC<a href="module-3---lab-practical.html#deepec" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><a href="https://quay.io/repository/hallamlab/external_deepec?tab=tags" class="uri">https://quay.io/repository/hallamlab/external_deepec?tab=tags</a></p>
<div class="sourceCode" id="cb77"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb77-1"><a href="module-3---lab-practical.html#cb77-1"></a><span class="co"># docker://quay.io/hallamlab/external_deepec:0.4.1</span></span>
<span id="cb77-2"><a href="module-3---lab-practical.html#cb77-2"></a>deepec_sif <span class="op">=</span> LIB<span class="op">/</span><span class="st">&quot;deepec.sif&quot;</span></span></code></pre></div>
<p>The DeepEC model uses convolutional neural networks, a deep learning architecture that is not far from HMM approach of Kofamscan.
It too scans the protein sequence and attempts to predict function based on the “topology” or motif structure of the amino acid sequences.
However, the model is able to make additional abstract decisions after the results of the initial scan in the form of additional model layers,
further differentiating it from the sequence similarity approach [1]. As the name suggests, deepEC predicts enzyme commission (EC) numbers,
which can be linked into the KEGG database, among others. It is straightforward to run.</p>
<ol style="list-style-type: decimal">
<li>Ryu JY, Kim HU, Lee SY. Deep learning enables high-quality and high-throughput prediction of enzyme commission numbers. Proc Natl Acad Sci. 2019;116(28):13996–4001. <a href="https://doi.org/10.1073/pnas.1821905116" class="uri">https://doi.org/10.1073/pnas.1821905116</a></li>
</ol>
<div class="sourceCode" id="cb78"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb78-1"><a href="module-3---lab-practical.html#cb78-1"></a>Shell(<span class="ss">f&quot;&quot;&quot;</span></span>
<span id="cb78-2"><a href="module-3---lab-practical.html#cb78-2"></a><span class="ss">apptainer exec </span><span class="sc">{</span>deepec_sif<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb78-3"><a href="module-3---lab-practical.html#cb78-3"></a><span class="ss">    deepec </span><span class="op">\</span></span>
<span id="cb78-4"><a href="module-3---lab-practical.html#cb78-4"></a><span class="ss">        -p </span><span class="sc">{</span>CPUS<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb78-5"><a href="module-3---lab-practical.html#cb78-5"></a><span class="ss">        -i </span><span class="sc">{</span>prodigal_aa<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb78-6"><a href="module-3---lab-practical.html#cb78-6"></a><span class="ss">        -o ./deepec</span></span>
<span id="cb78-7"><a href="module-3---lab-practical.html#cb78-7"></a><span class="ss">&quot;&quot;&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb79"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb79-1"><a href="module-3---lab-practical.html#cb79-1"></a>dfec <span class="op">=</span> pd.read_csv(<span class="st">&quot;./deepec/DeepEC_Result.txt&quot;</span>, sep<span class="op">=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</span>
<span id="cb79-2"><a href="module-3---lab-practical.html#cb79-2"></a><span class="bu">print</span>(dfec.shape)</span>
<span id="cb79-3"><a href="module-3---lab-practical.html#cb79-3"></a>dfec.head(<span class="dv">2</span>)</span></code></pre></div>
<details>
<summary>
How many ORFs did deepEC assign EC numbers to?
</summary>
<p>87 out of 7176</p>
<hr />
</details>
</div>
<div id="deepfri" class="section level3 hasAnchor">
<h3>deepFRI<a href="module-3---lab-practical.html#deepfri" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><a href="https://quay.io/repository/hallamlab/external_deepfri?tab=tags" class="uri">https://quay.io/repository/hallamlab/external_deepfri?tab=tags</a></p>
<div class="sourceCode" id="cb80"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb80-1"><a href="module-3---lab-practical.html#cb80-1"></a><span class="co"># docker://quay.io/hallamlab/external_deepfri:1.0.1</span></span>
<span id="cb80-2"><a href="module-3---lab-practical.html#cb80-2"></a>deepfri_sif <span class="op">=</span> LIB<span class="op">/</span><span class="st">&quot;deepfri.sif&quot;</span></span></code></pre></div>
<p>So deep learning models are able to consider the protein sequence at a higher-level of abstraction, but what is it abstracting to?
We often don’t know the mental model of protein function that these approaches have constucted for themselves, but deepFRI provides
and example that can ease us into the deep learning world. DeepFRI predicts function by first examining the potential structure
of a protein sequence by predicting the pairwise distances between all amino acids in the form of a “contact map” [1]. The final output
are Gene ontology (GO) numbers, which are an alternative to KEGG more common in medical settings.</p>
<ol style="list-style-type: decimal">
<li>Gligorijević V, Renfrew PD, Kosciolek T, Leman JK, Berenberg D, Vatanen T, et al. Structure-based protein function prediction using graph convolutional networks. Nat Commun. 2021;12(1):3168. <a href="https://doi.org/10.1038/s41467-021-23303-9" class="uri">https://doi.org/10.1038/s41467-021-23303-9</a></li>
</ol>
<p>DeepFRI is quite tempermental. We must remove the <code>*</code> character indicating stop codons from the prodigal output.</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb81-1"><a href="module-3---lab-practical.html#cb81-1"></a>prodigal_aa_no_stop <span class="op">=</span> Path(<span class="st">&quot;./prodigal/orfs.no_stop.faa&quot;</span>)</span>
<span id="cb81-2"><a href="module-3---lab-practical.html#cb81-2"></a><span class="cf">with</span> <span class="bu">open</span>(prodigal_aa_no_stop, <span class="st">&quot;w&quot;</span>) <span class="im">as</span> f:</span>
<span id="cb81-3"><a href="module-3---lab-practical.html#cb81-3"></a>    <span class="cf">for</span> e <span class="kw">in</span> SeqIO.parse(prodigal_aa, <span class="st">&quot;fasta&quot;</span>):</span>
<span id="cb81-4"><a href="module-3---lab-practical.html#cb81-4"></a>        <span class="cf">if</span> e.seq[<span class="op">-</span><span class="dv">1</span>] <span class="op">==</span> <span class="st">&quot;*&quot;</span>:</span>
<span id="cb81-5"><a href="module-3---lab-practical.html#cb81-5"></a>            e.seq <span class="op">=</span> e.seq[:<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb81-6"><a href="module-3---lab-practical.html#cb81-6"></a>        SeqIO.write(e, f, <span class="st">&quot;fasta&quot;</span>)</span></code></pre></div>
<p>We must also create and run it within a dedicated output folder, while using absolute paths.
As it runs, it will dump files into the current directory despite us telling it otherwise.
Try with the reduced set of hypotheticals (<code>{LIB}/"outputs/bakta.full/assembly.hypotheticals.faa"</code>)
or use the pre-computed results at (<code>{LIB}/"outputs/deepfri/</code>).</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb82-1"><a href="module-3---lab-practical.html#cb82-1"></a>prodigal_aa_no_stop <span class="op">=</span> Path(<span class="st">&quot;./prodigal/orfs.no_stop.faa&quot;</span>)</span>
<span id="cb82-2"><a href="module-3---lab-practical.html#cb82-2"></a>deepfri_out <span class="op">=</span> Path(<span class="st">&quot;./deepfri&quot;</span>)</span>
<span id="cb82-3"><a href="module-3---lab-practical.html#cb82-3"></a>Shell(<span class="ss">f&quot;&quot;&quot;</span></span>
<span id="cb82-4"><a href="module-3---lab-practical.html#cb82-4"></a><span class="ss">mkdir -p </span><span class="sc">{</span>deepfri_out<span class="sc">}</span></span>
<span id="cb82-5"><a href="module-3---lab-practical.html#cb82-5"></a><span class="ss">cp </span><span class="sc">{</span>prodigal_aa_no_stop<span class="sc">}</span><span class="ss"> </span><span class="sc">{</span>deepfri_out<span class="sc">}</span><span class="ss">/</span></span>
<span id="cb82-6"><a href="module-3---lab-practical.html#cb82-6"></a><span class="ss">cd </span><span class="sc">{</span>deepfri_out<span class="sc">}</span></span>
<span id="cb82-7"><a href="module-3---lab-practical.html#cb82-7"></a><span class="ss">apptainer exec -B ./:/ws -W /ws ../</span><span class="sc">{</span>deepfri_sif<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb82-8"><a href="module-3---lab-practical.html#cb82-8"></a><span class="ss">    deepfri </span><span class="op">\</span></span>
<span id="cb82-9"><a href="module-3---lab-practical.html#cb82-9"></a><span class="ss">        --ont mf bp ec </span><span class="op">\</span></span>
<span id="cb82-10"><a href="module-3---lab-practical.html#cb82-10"></a><span class="ss">        --fasta_fn /ws/</span><span class="sc">{</span>prodigal_aa_no_stop<span class="sc">.</span>name<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb82-11"><a href="module-3---lab-practical.html#cb82-11"></a><span class="ss">        --output_fn_prefix /ws/chicken</span></span>
<span id="cb82-12"><a href="module-3---lab-practical.html#cb82-12"></a><span class="ss">&quot;&quot;&quot;</span>)</span></code></pre></div>
<p>EC:</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb83-1"><a href="module-3---lab-practical.html#cb83-1"></a>df <span class="op">=</span> pd.read_csv(LIB<span class="op">/</span><span class="st">&quot;outputs/deepfri/chicken_EC_predictions.csv&quot;</span>, comment<span class="op">=</span><span class="st">&quot;#&quot;</span>)</span>
<span id="cb83-2"><a href="module-3---lab-practical.html#cb83-2"></a><span class="bu">print</span>(df.shape)</span>
<span id="cb83-3"><a href="module-3---lab-practical.html#cb83-3"></a>df.head(<span class="dv">2</span>)</span></code></pre></div>
<p>GO molecular function</p>
<div class="sourceCode" id="cb84"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb84-1"><a href="module-3---lab-practical.html#cb84-1"></a>df <span class="op">=</span> pd.read_csv(LIB<span class="op">/</span><span class="st">&quot;outputs/deepfri/chicken_MF_predictions.csv&quot;</span>, comment<span class="op">=</span><span class="st">&quot;#&quot;</span>)</span>
<span id="cb84-2"><a href="module-3---lab-practical.html#cb84-2"></a><span class="bu">print</span>(df.shape)</span>
<span id="cb84-3"><a href="module-3---lab-practical.html#cb84-3"></a>df.head(<span class="dv">2</span>)</span></code></pre></div>
<p>GO biological process</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb85-1"><a href="module-3---lab-practical.html#cb85-1"></a>df <span class="op">=</span> pd.read_csv(LIB<span class="op">/</span><span class="st">&quot;outputs/deepfri/chicken_BP_predictions.csv&quot;</span>, comment<span class="op">=</span><span class="st">&quot;#&quot;</span>)</span>
<span id="cb85-2"><a href="module-3---lab-practical.html#cb85-2"></a><span class="bu">print</span>(df.shape)</span>
<span id="cb85-3"><a href="module-3---lab-practical.html#cb85-3"></a>df.head(<span class="dv">2</span>)</span></code></pre></div>
</div>
<div id="proteinbert" class="section level3 hasAnchor">
<h3>proteinBERT<a href="module-3---lab-practical.html#proteinbert" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><a href="https://quay.io/repository/hallamlab/external_proteinbert?tab=tags" class="uri">https://quay.io/repository/hallamlab/external_proteinbert?tab=tags</a></p>
<div class="sourceCode" id="cb86"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb86-1"><a href="module-3---lab-practical.html#cb86-1"></a><span class="co"># docker://quay.io/hallamlab/external_proteinbert:2024.03.28</span></span>
<span id="cb86-2"><a href="module-3---lab-practical.html#cb86-2"></a>proteinbert_sif <span class="op">=</span> LIB<span class="op">/</span><span class="st">&quot;proteinbert.sif&quot;</span></span></code></pre></div>
<p>ProteinBERT “reads” protein sequences like reading text. The core of its architecture is an attention mechanism that is
part of what’s called a “transformer”, often used in natural language processing [1, 2]. Residue context is considered and pairwise
importance is predicted to enable the model to “focus” on what it thinks are important motifs. The pairwise attention could be
spatial distance, as in deepFRI, but is not strictly limited to this interpretation, hence increasing the abstraction capabilities
of the model. The core transformer architecture is used in current large language models like ChatGPT. The only difference is that
ChatGPT is much, much larger.</p>
<ol style="list-style-type: decimal">
<li>Vaswani A, Shazeer N, Parmar N, Uszkoreit J, Jones L, Gomez AN, et al. Attention is all you need. Curran Associates Inc.; 2017. (NIPS’17).</li>
<li>Brandes N, Ofer D, Peleg Y, Rappoport N, Linial M. ProteinBERT: a universal deep-learning model of protein sequence and function. Bioinformatics. 2022;38(8):2102–10. <a href="https://doi.org/10.1093/bioinformatics/btac020" class="uri">https://doi.org/10.1093/bioinformatics/btac020</a></li>
</ol>
<p>Since the model did not have friendly interface, a CLI was constructed courtesy of the Hallam Lab.</p>
<div class="sourceCode" id="cb87"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb87-1"><a href="module-3---lab-practical.html#cb87-1"></a>apptainer <span class="bu">exec</span> {proteinbert_sif} <span class="op">\</span></span>
<span id="cb87-2"><a href="module-3---lab-practical.html#cb87-2"></a>    pbert run <span class="op">\</span></span>
<span id="cb87-3"><a href="module-3---lab-practical.html#cb87-3"></a>        <span class="op">--</span>threads {CPUS} <span class="op">\</span></span>
<span id="cb87-4"><a href="module-3---lab-practical.html#cb87-4"></a>        <span class="op">-</span>i {prodigal_aa} <span class="op">\</span></span>
<span id="cb87-5"><a href="module-3---lab-practical.html#cb87-5"></a>        <span class="op">-</span>o protein_bert</span></code></pre></div>
<p>The output of proteinBERT is an array of numbers for each ORF, representing their predicted function.
These numbers can be used as coordinates in proteinBERT’s mental model.
We say each ORF is embedded in proteinBERT’s functional space. The euclidean distance between each ORF’s
embedding represents predicted functional similarity.</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb88-1"><a href="module-3---lab-practical.html#cb88-1"></a>df_index <span class="op">=</span> pd.read_csv(<span class="st">&quot;./protein_bert/orfs.csv&quot;</span>)</span>
<span id="cb88-2"><a href="module-3---lab-practical.html#cb88-2"></a><span class="bu">print</span>(df_index.shape)</span>
<span id="cb88-3"><a href="module-3---lab-practical.html#cb88-3"></a>df_index.head(<span class="dv">2</span>)</span></code></pre></div>
<p>ProteinBERT has 6 transformer layers, each of which reads then paraphrases the aminoacid sequence for the next.
This information is stored in a 512-dimensional vector for each ORF, so the total embedding matrix is
<code>N_orfs x (6x512)</code>.</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb89-1"><a href="module-3---lab-practical.html#cb89-1"></a>mats <span class="op">=</span> []</span>
<span id="cb89-2"><a href="module-3---lab-practical.html#cb89-2"></a><span class="cf">for</span> batch <span class="kw">in</span> <span class="bu">sorted</span>(df_index.batch.unique()):</span>
<span id="cb89-3"><a href="module-3---lab-practical.html#cb89-3"></a>    mats.append(np.load(<span class="ss">f&quot;./protein_bert/orfs.</span><span class="sc">{</span>batch<span class="sc">}</span><span class="ss">.embedding.npy&quot;</span>))</span>
<span id="cb89-4"><a href="module-3---lab-practical.html#cb89-4"></a>mat <span class="op">=</span> np.vstack(mats)</span>
<span id="cb89-5"><a href="module-3---lab-practical.html#cb89-5"></a>mat.shape</span></code></pre></div>
<p>Examining the standard deviation of the embeddings for each layers reveals that the last layer has the most to say,
so we will use it to reduce needed compute.</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb90-1"><a href="module-3---lab-practical.html#cb90-1"></a>fig <span class="op">=</span> BaseFigure()</span>
<span id="cb90-2"><a href="module-3---lab-practical.html#cb90-2"></a>fig.add_trace(</span>
<span id="cb90-3"><a href="module-3---lab-practical.html#cb90-3"></a>    go.Scatter(</span>
<span id="cb90-4"><a href="module-3---lab-practical.html#cb90-4"></a>        x<span class="op">=</span>np.arange(mat.shape[<span class="dv">0</span>]),</span>
<span id="cb90-5"><a href="module-3---lab-practical.html#cb90-5"></a>        y<span class="op">=</span>mat.std(axis<span class="op">=</span><span class="dv">0</span>),</span>
<span id="cb90-6"><a href="module-3---lab-practical.html#cb90-6"></a>        mode<span class="op">=</span><span class="st">&quot;markers&quot;</span>,</span>
<span id="cb90-7"><a href="module-3---lab-practical.html#cb90-7"></a>        marker<span class="op">=</span><span class="bu">dict</span>(size<span class="op">=</span><span class="dv">2</span>),</span>
<span id="cb90-8"><a href="module-3---lab-practical.html#cb90-8"></a>    )</span>
<span id="cb90-9"><a href="module-3---lab-practical.html#cb90-9"></a>)</span>
<span id="cb90-10"><a href="module-3---lab-practical.html#cb90-10"></a>fig <span class="op">=</span> ApplyTemplate(fig, default_xaxis<span class="op">=</span><span class="bu">dict</span>(dtick<span class="op">=</span><span class="dv">512</span>))</span>
<span id="cb90-11"><a href="module-3---lab-practical.html#cb90-11"></a>fig.show()</span></code></pre></div>
<div class="sourceCode" id="cb91"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb91-1"><a href="module-3---lab-practical.html#cb91-1"></a>mat_last <span class="op">=</span> mat[:, <span class="op">-</span><span class="dv">512</span>:]</span>
<span id="cb91-2"><a href="module-3---lab-practical.html#cb91-2"></a>mat_last.shape</span></code></pre></div>
<p>Let’s load the kofamscan results to provide KEGG landmarks for these embeddings.</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb92-1"><a href="module-3---lab-practical.html#cb92-1"></a>LIB <span class="op">=</span> Path(<span class="st">&quot;./lib&quot;</span>)</span>
<span id="cb92-2"><a href="module-3---lab-practical.html#cb92-2"></a>kofamscan_raw <span class="op">=</span> Path(<span class="st">&quot;./kofamscan/kos.out&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb93"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb93-1"><a href="module-3---lab-practical.html#cb93-1"></a><span class="im">from</span> local.models.kegg_orthology <span class="im">import</span> ParseKofamScanResults</span>
<span id="cb93-2"><a href="module-3---lab-practical.html#cb93-2"></a></span>
<span id="cb93-3"><a href="module-3---lab-practical.html#cb93-3"></a>ko_model <span class="op">=</span> ParseKofamScanResults(</span>
<span id="cb93-4"><a href="module-3---lab-practical.html#cb93-4"></a>    kofamscan_raw,</span>
<span id="cb93-5"><a href="module-3---lab-practical.html#cb93-5"></a>    LIB<span class="op">/</span><span class="st">&quot;kofamscan_db/api_kegg.db&quot;</span>,</span>
<span id="cb93-6"><a href="module-3---lab-practical.html#cb93-6"></a>    LIB<span class="op">/</span><span class="st">&quot;kofamscan_db/brite.json&quot;</span>,</span>
<span id="cb93-7"><a href="module-3---lab-practical.html#cb93-7"></a>)</span></code></pre></div>
<div class="sourceCode" id="cb94"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb94-1"><a href="module-3---lab-practical.html#cb94-1"></a>df_kofam <span class="op">=</span> ko_model.df</span>
<span id="cb94-2"><a href="module-3---lab-practical.html#cb94-2"></a>df_kofam <span class="op">=</span> df_kofam[df_kofam[<span class="st">&quot;score&quot;</span>]<span class="op">&gt;</span>df_kofam[<span class="st">&quot;hmm_threshold&quot;</span>]]</span>
<span id="cb94-3"><a href="module-3---lab-practical.html#cb94-3"></a><span class="bu">print</span>(df_kofam.shape)</span>
<span id="cb94-4"><a href="module-3---lab-practical.html#cb94-4"></a>df_kofam.head(<span class="dv">2</span>)</span></code></pre></div>
<div class="sourceCode" id="cb95"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb95-1"><a href="module-3---lab-practical.html#cb95-1"></a>categories <span class="op">=</span> {}</span>
<span id="cb95-2"><a href="module-3---lab-practical.html#cb95-2"></a><span class="cf">for</span> _, row <span class="kw">in</span> df_kofam.iterrows():</span>
<span id="cb95-3"><a href="module-3---lab-practical.html#cb95-3"></a>    ko <span class="op">=</span> row[<span class="st">&quot;ko&quot;</span>]</span>
<span id="cb95-4"><a href="module-3---lab-practical.html#cb95-4"></a>    orf <span class="op">=</span> row[<span class="st">&quot;orf&quot;</span>]</span>
<span id="cb95-5"><a href="module-3---lab-practical.html#cb95-5"></a>    <span class="cf">for</span> brite2, brite1 <span class="kw">in</span> ko_model.GetCategories(ko):</span>
<span id="cb95-6"><a href="module-3---lab-practical.html#cb95-6"></a>        categories[brite2] <span class="op">=</span> categories.get(brite2, <span class="bu">set</span>()) <span class="op">|</span> {orf}</span>
<span id="cb95-7"><a href="module-3---lab-practical.html#cb95-7"></a></span>
<span id="cb95-8"><a href="module-3---lab-practical.html#cb95-8"></a>categories <span class="op">=</span> <span class="bu">sorted</span>(categories.items(), key<span class="op">=</span><span class="kw">lambda</span> x: <span class="bu">len</span>(x[<span class="dv">1</span>]), reverse<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb95-9"><a href="module-3---lab-practical.html#cb95-9"></a><span class="bu">len</span>(categories)</span></code></pre></div>
<p>This assigns each ORF a category based on the BRITE hierarchy.</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb96-1"><a href="module-3---lab-practical.html#cb96-1"></a>orf2cat <span class="op">=</span> {}</span>
<span id="cb96-2"><a href="module-3---lab-practical.html#cb96-2"></a><span class="cf">for</span> brite2, orfs <span class="kw">in</span> categories:</span>
<span id="cb96-3"><a href="module-3---lab-practical.html#cb96-3"></a>    <span class="cf">for</span> orf <span class="kw">in</span> orfs:</span>
<span id="cb96-4"><a href="module-3---lab-practical.html#cb96-4"></a>        <span class="cf">if</span> orf <span class="kw">in</span> orf2cat: <span class="cf">continue</span></span>
<span id="cb96-5"><a href="module-3---lab-practical.html#cb96-5"></a>        orf2cat[orf] <span class="op">=</span> brite2</span>
<span id="cb96-6"><a href="module-3---lab-practical.html#cb96-6"></a>cat2i <span class="op">=</span> {cat: i<span class="op">+</span><span class="dv">1</span> <span class="cf">for</span> i, (cat, _) <span class="kw">in</span> <span class="bu">enumerate</span>(categories)}</span>
<span id="cb96-7"><a href="module-3---lab-practical.html#cb96-7"></a></span>
<span id="cb96-8"><a href="module-3---lab-practical.html#cb96-8"></a>y_categories <span class="op">=</span> []</span>
<span id="cb96-9"><a href="module-3---lab-practical.html#cb96-9"></a><span class="cf">for</span> _, row <span class="kw">in</span> df_index.iterrows():</span>
<span id="cb96-10"><a href="module-3---lab-practical.html#cb96-10"></a>    orf <span class="op">=</span> row[<span class="st">&quot;id&quot;</span>]</span>
<span id="cb96-11"><a href="module-3---lab-practical.html#cb96-11"></a>    brite2 <span class="op">=</span> orf2cat.get(orf, <span class="st">&quot;None&quot;</span>)</span>
<span id="cb96-12"><a href="module-3---lab-practical.html#cb96-12"></a>    i <span class="op">=</span> cat2i.get(brite2, <span class="dv">0</span>)</span>
<span id="cb96-13"><a href="module-3---lab-practical.html#cb96-13"></a>    y_categories.append(i)</span></code></pre></div>
<p>We can use UMAP to reduce the 512 dimensional embeddings to 2 dimensions for visualization
and use the BRITE categories as a guide.</p>
<div class="sourceCode" id="cb97"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb97-1"><a href="module-3---lab-practical.html#cb97-1"></a><span class="im">from</span> umap <span class="im">import</span> UMAP</span>
<span id="cb97-2"><a href="module-3---lab-practical.html#cb97-2"></a></span>
<span id="cb97-3"><a href="module-3---lab-practical.html#cb97-3"></a>model <span class="op">=</span> UMAP(</span>
<span id="cb97-4"><a href="module-3---lab-practical.html#cb97-4"></a>    n_components<span class="op">=</span><span class="dv">2</span>, <span class="co"># reduce to x, y</span></span>
<span id="cb97-5"><a href="module-3---lab-practical.html#cb97-5"></a>    spread<span class="op">=</span><span class="dv">5</span>,       <span class="co"># spread out the points</span></span>
<span id="cb97-6"><a href="module-3---lab-practical.html#cb97-6"></a>    min_dist<span class="op">=</span><span class="fl">0.1</span>,   <span class="co"># reduce the maximum density of points</span></span>
<span id="cb97-7"><a href="module-3---lab-practical.html#cb97-7"></a>    metric<span class="op">=</span><span class="st">&quot;cosine&quot;</span>,</span>
<span id="cb97-8"><a href="module-3---lab-practical.html#cb97-8"></a>)</span>
<span id="cb97-9"><a href="module-3---lab-practical.html#cb97-9"></a></span>
<span id="cb97-10"><a href="module-3---lab-practical.html#cb97-10"></a>emb <span class="op">=</span> model.fit_transform(mat_last, y<span class="op">=</span>y_categories)</span></code></pre></div>
<p>Let’s have a look.</p>
<div class="sourceCode" id="cb98"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb98-1"><a href="module-3---lab-practical.html#cb98-1"></a><span class="im">from</span> local.figures.template <span class="im">import</span> BaseFigure, ApplyTemplate, go</span>
<span id="cb98-2"><a href="module-3---lab-practical.html#cb98-2"></a><span class="im">from</span> local.figures.colors <span class="im">import</span> Palettes, COLORS</span>
<span id="cb98-3"><a href="module-3---lab-practical.html#cb98-3"></a></span>
<span id="cb98-4"><a href="module-3---lab-practical.html#cb98-4"></a>fig <span class="op">=</span> BaseFigure()</span>
<span id="cb98-5"><a href="module-3---lab-practical.html#cb98-5"></a></span>
<span id="cb98-6"><a href="module-3---lab-practical.html#cb98-6"></a>K <span class="op">=</span> <span class="dv">10</span></span>
<span id="cb98-7"><a href="module-3---lab-practical.html#cb98-7"></a>seen <span class="op">=</span> <span class="bu">set</span>()</span>
<span id="cb98-8"><a href="module-3---lab-practical.html#cb98-8"></a>todo <span class="op">=</span> []</span>
<span id="cb98-9"><a href="module-3---lab-practical.html#cb98-9"></a><span class="cf">for</span> brite2, _ <span class="kw">in</span> categories[:K]:</span>
<span id="cb98-10"><a href="module-3---lab-practical.html#cb98-10"></a>    _filter <span class="op">=</span> [orf2cat.get(orf) <span class="op">==</span> brite2 <span class="cf">for</span> orf <span class="kw">in</span> df_index[<span class="st">&quot;id&quot;</span>]]</span>
<span id="cb98-11"><a href="module-3---lab-practical.html#cb98-11"></a>    seen <span class="op">|=</span> <span class="bu">set</span>(orfs)</span>
<span id="cb98-12"><a href="module-3---lab-practical.html#cb98-12"></a>    todo.append((brite2, _filter))</span>
<span id="cb98-13"><a href="module-3---lab-practical.html#cb98-13"></a>_palette <span class="op">=</span> Palettes.PLOTLY</span>
<span id="cb98-14"><a href="module-3---lab-practical.html#cb98-14"></a><span class="cf">for</span> i, (name, _filter) <span class="kw">in</span> <span class="bu">enumerate</span>(todo):</span>
<span id="cb98-15"><a href="module-3---lab-practical.html#cb98-15"></a>    fig.add_trace(</span>
<span id="cb98-16"><a href="module-3---lab-practical.html#cb98-16"></a>        go.Scatter(</span>
<span id="cb98-17"><a href="module-3---lab-practical.html#cb98-17"></a>            name<span class="op">=</span>ko_model.GetNodeInfo(name)[<span class="dv">1</span>],</span>
<span id="cb98-18"><a href="module-3---lab-practical.html#cb98-18"></a>            x<span class="op">=</span>emb[_filter, <span class="dv">0</span>],</span>
<span id="cb98-19"><a href="module-3---lab-practical.html#cb98-19"></a>            y<span class="op">=</span>emb[_filter, <span class="dv">1</span>],</span>
<span id="cb98-20"><a href="module-3---lab-practical.html#cb98-20"></a>            text<span class="op">=</span>df_index[_filter][<span class="st">&quot;id&quot;</span>],</span>
<span id="cb98-21"><a href="module-3---lab-practical.html#cb98-21"></a>            mode<span class="op">=</span><span class="st">&quot;markers&quot;</span>,</span>
<span id="cb98-22"><a href="module-3---lab-practical.html#cb98-22"></a>            marker<span class="op">=</span><span class="bu">dict</span>(</span>
<span id="cb98-23"><a href="module-3---lab-practical.html#cb98-23"></a>                size<span class="op">=</span><span class="dv">10</span>,</span>
<span id="cb98-24"><a href="module-3---lab-practical.html#cb98-24"></a>                color<span class="op">=</span>COLORS.TRANSPARENT,</span>
<span id="cb98-25"><a href="module-3---lab-practical.html#cb98-25"></a>                line<span class="op">=</span><span class="bu">dict</span>(</span>
<span id="cb98-26"><a href="module-3---lab-practical.html#cb98-26"></a>                    width<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb98-27"><a href="module-3---lab-practical.html#cb98-27"></a>                    color<span class="op">=</span>_palette[i<span class="op">%</span><span class="bu">len</span>(_palette)].color_value,</span>
<span id="cb98-28"><a href="module-3---lab-practical.html#cb98-28"></a>                ),</span>
<span id="cb98-29"><a href="module-3---lab-practical.html#cb98-29"></a>            ),</span>
<span id="cb98-30"><a href="module-3---lab-practical.html#cb98-30"></a>        )</span>
<span id="cb98-31"><a href="module-3---lab-practical.html#cb98-31"></a>    )</span>
<span id="cb98-32"><a href="module-3---lab-practical.html#cb98-32"></a></span>
<span id="cb98-33"><a href="module-3---lab-practical.html#cb98-33"></a><span class="co"># plot rest as gray dots</span></span>
<span id="cb98-34"><a href="module-3---lab-practical.html#cb98-34"></a>_filter <span class="op">=</span> <span class="op">~</span>df_index[<span class="st">&quot;id&quot;</span>].isin(seen)</span>
<span id="cb98-35"><a href="module-3---lab-practical.html#cb98-35"></a>fig.add_trace(</span>
<span id="cb98-36"><a href="module-3---lab-practical.html#cb98-36"></a>    go.Scatter(</span>
<span id="cb98-37"><a href="module-3---lab-practical.html#cb98-37"></a>        name<span class="op">=</span><span class="st">&quot;other&quot;</span>,</span>
<span id="cb98-38"><a href="module-3---lab-practical.html#cb98-38"></a>        x<span class="op">=</span>emb[_filter, <span class="dv">0</span>],</span>
<span id="cb98-39"><a href="module-3---lab-practical.html#cb98-39"></a>        y<span class="op">=</span>emb[_filter, <span class="dv">1</span>],</span>
<span id="cb98-40"><a href="module-3---lab-practical.html#cb98-40"></a>        text<span class="op">=</span>df_index[_filter][<span class="st">&quot;id&quot;</span>],</span>
<span id="cb98-41"><a href="module-3---lab-practical.html#cb98-41"></a>        mode<span class="op">=</span><span class="st">&quot;markers&quot;</span>,</span>
<span id="cb98-42"><a href="module-3---lab-practical.html#cb98-42"></a>        marker<span class="op">=</span><span class="bu">dict</span>(</span>
<span id="cb98-43"><a href="module-3---lab-practical.html#cb98-43"></a>            size<span class="op">=</span><span class="dv">5</span>, opacity<span class="op">=</span><span class="fl">0.1</span>,</span>
<span id="cb98-44"><a href="module-3---lab-practical.html#cb98-44"></a>            color<span class="op">=</span>COLORS.GRAY,</span>
<span id="cb98-45"><a href="module-3---lab-practical.html#cb98-45"></a>        ),</span>
<span id="cb98-46"><a href="module-3---lab-practical.html#cb98-46"></a>    )</span>
<span id="cb98-47"><a href="module-3---lab-practical.html#cb98-47"></a>)</span>
<span id="cb98-48"><a href="module-3---lab-practical.html#cb98-48"></a>no_axis <span class="op">=</span> <span class="bu">dict</span>(linecolor<span class="op">=</span>COLORS.TRANSPARENT, ticks<span class="op">=</span><span class="va">None</span>, showticklabels<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb98-49"><a href="module-3---lab-practical.html#cb98-49"></a>fig <span class="op">=</span> ApplyTemplate(</span>
<span id="cb98-50"><a href="module-3---lab-practical.html#cb98-50"></a>    fig,</span>
<span id="cb98-51"><a href="module-3---lab-practical.html#cb98-51"></a>    default_xaxis<span class="op">=</span>no_axis, default_yaxis<span class="op">=</span>no_axis,</span>
<span id="cb98-52"><a href="module-3---lab-practical.html#cb98-52"></a>    layout<span class="op">=</span><span class="bu">dict</span>(</span>
<span id="cb98-53"><a href="module-3---lab-practical.html#cb98-53"></a>        width<span class="op">=</span><span class="dv">1000</span>,</span>
<span id="cb98-54"><a href="module-3---lab-practical.html#cb98-54"></a>        height<span class="op">=</span><span class="dv">700</span>,</span>
<span id="cb98-55"><a href="module-3---lab-practical.html#cb98-55"></a>    ),</span>
<span id="cb98-56"><a href="module-3---lab-practical.html#cb98-56"></a>)</span>
<span id="cb98-57"><a href="module-3---lab-practical.html#cb98-57"></a>fig.show()</span></code></pre></div>
<details>
<summary>
What are the gray dots?
</summary>
<p>These are ORFs not in the top K categories, including ORFs of unknown function.</p>
<hr />
</details>
</div>
</div>
<div id="gapseq" class="section level2 hasAnchor">
<h2>gapseq<a href="module-3---lab-practical.html#gapseq" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><a href="https://quay.io/repository/biocontainers/gapseq?tab=tags" class="uri">https://quay.io/repository/biocontainers/gapseq?tab=tags</a></p>
<div class="sourceCode" id="cb99"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb99-1"><a href="module-3---lab-practical.html#cb99-1"></a><span class="co"># docker://quay.io/biocontainers/gapseq:1.4.0--h9ee0642_1</span></span>
<span id="cb99-2"><a href="module-3---lab-practical.html#cb99-2"></a>gapseqt_sif <span class="op">=</span> LIB<span class="op">/</span><span class="st">&quot;gapseq.sif&quot;</span></span></code></pre></div>
<p>So far, we have used various tools to predict the locations of genes and their functions in our metagenomic assembly.
A portion of these functions were enzymatic reactions which convert between chemical compounds called metabolites.
Others facilitate the transport of metabolites across membranes. Imagine reactions and transporters to be the edges of a network,
chained together by common metabolites. The combined network would model the metabolic capacity encoded in the examined genetic sequences,
drawing out possible pathways by which compounds are chemically transformed. Models from single genomes can indicate viability
by possesing all necessary pathways to produce the constituents of biomass (nucleotides, amino acids, cofactors, lipids, etc.).
Prediction errors that cause reactions to be missing create gaps in pathways, removing metabolic capacity, and potentially rendering
the model non-viable. <code>gapseq</code> is a tool for filling these gaps from reference metabolic models by weighing candidate reactions
against sequence evidence [1]. We will use gapseq to generate a model for each of the 2 MAGs by assuming viability on M9 minimal media.</p>
<ol style="list-style-type: decimal">
<li>Zimmermann J, Kaleta C, Waschina S. gapseq: informed prediction of bacterial metabolic pathways and reconstruction of accurate metabolic models. Genome Biol. 2021;22(1):81. <a href="https://doi.org/10.1186/s13059-021-02295-1" class="uri">https://doi.org/10.1186/s13059-021-02295-1</a></li>
</ol>
<details>
<summary>
Can we run gapseq on the entire metagenomic assembly?
</summary>
<p>This would be a crude model of the community’s metabolic capacity and can be used to detect possible pathways distributed across
different members. A major source of error would be the lack of compartmentalization. The fact that a metabolite would have to be
transported out of the cytosol of one cell and into the cytosol of another cell is not considered.</p>
<hr />
</details>
<p>First, we separate the ORFs by MAG.</p>
<div class="sourceCode" id="cb100"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb100-1"><a href="module-3---lab-practical.html#cb100-1"></a>mag1_orfs, mag2_orfs <span class="op">=</span> [], []</span>
<span id="cb100-2"><a href="module-3---lab-practical.html#cb100-2"></a><span class="cf">for</span> e <span class="kw">in</span> SeqIO.parse(prodigal_aa, <span class="st">&quot;fasta&quot;</span>):</span>
<span id="cb100-3"><a href="module-3---lab-practical.html#cb100-3"></a>    contig <span class="op">=</span> <span class="st">&quot;_&quot;</span>.join(e.<span class="bu">id</span>.split(<span class="st">&quot;_&quot;</span>)[:<span class="op">-</span><span class="dv">1</span>])</span>
<span id="cb100-4"><a href="module-3---lab-practical.html#cb100-4"></a>    <span class="cf">if</span> contig <span class="kw">in</span> MAG1:</span>
<span id="cb100-5"><a href="module-3---lab-practical.html#cb100-5"></a>        mag1_orfs.append(e)</span>
<span id="cb100-6"><a href="module-3---lab-practical.html#cb100-6"></a>    <span class="cf">elif</span> contig <span class="kw">in</span> MAG2:</span>
<span id="cb100-7"><a href="module-3---lab-practical.html#cb100-7"></a>        mag2_orfs.append(e)</span>
<span id="cb100-8"><a href="module-3---lab-practical.html#cb100-8"></a></span>
<span id="cb100-9"><a href="module-3---lab-practical.html#cb100-9"></a>mag1_aa, mag2_aa <span class="op">=</span> <span class="st">&quot;./prodigal/mag1.faa&quot;</span>, <span class="st">&quot;./prodigal/mag2.faa&quot;</span></span>
<span id="cb100-10"><a href="module-3---lab-practical.html#cb100-10"></a><span class="cf">with</span> <span class="bu">open</span>(mag1_aa, <span class="st">&quot;w&quot;</span>) <span class="im">as</span> f:</span>
<span id="cb100-11"><a href="module-3---lab-practical.html#cb100-11"></a>    SeqIO.write(mag1_orfs, f, <span class="st">&quot;fasta&quot;</span>)</span>
<span id="cb100-12"><a href="module-3---lab-practical.html#cb100-12"></a><span class="cf">with</span> <span class="bu">open</span>(mag2_aa, <span class="st">&quot;w&quot;</span>) <span class="im">as</span> f:</span>
<span id="cb100-13"><a href="module-3---lab-practical.html#cb100-13"></a>    SeqIO.write(mag2_orfs, f, <span class="st">&quot;fasta&quot;</span>)</span></code></pre></div>
<p>We will have to repeat the gapseq protocol for each MAG.</p>
<div class="sourceCode" id="cb101"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb101-1"><a href="module-3---lab-practical.html#cb101-1"></a>gapseq_mag <span class="op">=</span> <span class="st">&quot;mag1&quot;</span></span>
<span id="cb101-2"><a href="module-3---lab-practical.html#cb101-2"></a><span class="co"># gapseq_mag = &quot;mag2&quot; # repeat with mag2</span></span>
<span id="cb101-3"><a href="module-3---lab-practical.html#cb101-3"></a>aa_fasta <span class="op">=</span> <span class="ss">f&quot;./prodigal/</span><span class="sc">{</span>gapseq_mag<span class="sc">}</span><span class="ss">.faa&quot;</span></span>
<span id="cb101-4"><a href="module-3---lab-practical.html#cb101-4"></a>out_dir <span class="op">=</span> <span class="ss">f&quot;./gapseq/</span><span class="sc">{</span>gapseq_mag<span class="sc">}</span><span class="ss">&quot;</span></span></code></pre></div>
<p>Candidate reactions are searched for by gapseq <em>one pathway at a time</em>. This makes multiple (thousands)
of redundant calls to the alignment tool which makes the process incredibly slow. This is because gapseq
was intended to be an aid to manual curation and not a fully automated tool.
Pre-computed outputs are available at <code>{LIB}/outputs/gapseq</code>. A quick trial with only a single pathway
(<code>glycolysis</code>) is possible.</p>
<ul>
<li><code>-p keywords such as pathways or subsystems (for example amino,nucl,cofactor,carbo,polyamine)</code></li>
<li><code>-l Select the pathway database (MetaCyc, KEGG, SEED, all; default: metacyc,custom)</code></li>
<li><code>-K Number of threads for sequence alignments.</code></li>
<li><code>-O Force offline mode</code></li>
<li><code>-f Path to directory, where output files will be saved</code></li>
</ul>
<div class="sourceCode" id="cb102"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb102-1"><a href="module-3---lab-practical.html#cb102-1"></a>gapseq_p <span class="op">=</span> <span class="st">&quot;glycolysis&quot;</span></span>
<span id="cb102-2"><a href="module-3---lab-practical.html#cb102-2"></a><span class="co"># gapseq_p = &quot;all&quot; # takes about 4 hours!!</span></span>
<span id="cb102-3"><a href="module-3---lab-practical.html#cb102-3"></a>Shell(<span class="ss">f&quot;&quot;&quot;</span></span>
<span id="cb102-4"><a href="module-3---lab-practical.html#cb102-4"></a><span class="ss">mkdir -p </span><span class="sc">{</span>out_dir<span class="sc">}</span></span>
<span id="cb102-5"><a href="module-3---lab-practical.html#cb102-5"></a><span class="ss">apptainer exec </span><span class="sc">{</span>gapseq_sif<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb102-6"><a href="module-3---lab-practical.html#cb102-6"></a><span class="ss">    gapseq find -p </span><span class="sc">{</span>gapseq_p<span class="sc">}</span><span class="ss"> -l KEGG -K </span><span class="sc">{</span>CPU<span class="sc">}</span><span class="ss"> -O -f </span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss"> </span><span class="sc">{</span>aa_fasta<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb102-7"><a href="module-3---lab-practical.html#cb102-7"></a><span class="ss">        &gt; </span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/find.out 2&gt; </span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/find.err</span></span>
<span id="cb102-8"><a href="module-3---lab-practical.html#cb102-8"></a><span class="ss">&quot;&quot;&quot;</span>)</span></code></pre></div>
<details>
<summary>
<code>&gt;</code> and <code>2&gt;</code>
</summary>
<p><code>&gt;</code> redirects the standard output and <code>2&gt;</code> redirects the standard error to a files for logging.</p>
<hr />
</details>
<p>Transporters are faster, but still takes several minutes. Please use pre-computed results.</p>
<div class="sourceCode" id="cb103"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb103-1"><a href="module-3---lab-practical.html#cb103-1"></a>Shell(<span class="ss">f&quot;&quot;&quot;</span></span>
<span id="cb103-2"><a href="module-3---lab-practical.html#cb103-2"></a><span class="ss">mkdir -p </span><span class="sc">{</span>out_dir<span class="sc">}</span></span>
<span id="cb103-3"><a href="module-3---lab-practical.html#cb103-3"></a><span class="ss">apptainer exec </span><span class="sc">{</span>gapseq_sif<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb103-4"><a href="module-3---lab-practical.html#cb103-4"></a><span class="ss">    gapseq find-transport -K </span><span class="sc">{</span>CPU<span class="sc">}</span><span class="ss"> -f </span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss"> </span><span class="sc">{</span>aa_fasta<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb103-5"><a href="module-3---lab-practical.html#cb103-5"></a><span class="ss">        &gt; </span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/findtr.out 2&gt; </span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/findtr.err</span></span>
<span id="cb103-6"><a href="module-3---lab-practical.html#cb103-6"></a><span class="ss">&quot;&quot;&quot;</span>)</span></code></pre></div>
<p>Construct a draft model using candidate reactions, transporters, and pathways.</p>
<div class="sourceCode" id="cb104"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb104-1"><a href="module-3---lab-practical.html#cb104-1"></a>Shell(<span class="ss">f&quot;&quot;&quot;</span></span>
<span id="cb104-2"><a href="module-3---lab-practical.html#cb104-2"></a><span class="ss">apptainer exec </span><span class="sc">{</span>gapseq_sif<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb104-3"><a href="module-3---lab-practical.html#cb104-3"></a><span class="ss">    gapseq draft </span><span class="op">\</span></span>
<span id="cb104-4"><a href="module-3---lab-practical.html#cb104-4"></a><span class="ss">        -r </span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/</span><span class="sc">{</span>gapseq_mag<span class="sc">}</span><span class="ss">-</span><span class="sc">{</span>gapseq_p<span class="sc">}</span><span class="ss">-Reactions.tbl </span><span class="op">\</span></span>
<span id="cb104-5"><a href="module-3---lab-practical.html#cb104-5"></a><span class="ss">        -p </span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/</span><span class="sc">{</span>gapseq_mag<span class="sc">}</span><span class="ss">-</span><span class="sc">{</span>gapseq_p<span class="sc">}</span><span class="ss">-Pathways.tbl </span><span class="op">\</span></span>
<span id="cb104-6"><a href="module-3---lab-practical.html#cb104-6"></a><span class="ss">        -t </span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/</span><span class="sc">{</span>gapseq_mag<span class="sc">}</span><span class="ss">-Transporter.tbl </span><span class="op">\</span></span>
<span id="cb104-7"><a href="module-3---lab-practical.html#cb104-7"></a><span class="ss">        -c </span><span class="sc">{</span>aa_fasta<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb104-8"><a href="module-3---lab-practical.html#cb104-8"></a><span class="ss">        -f </span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb104-9"><a href="module-3---lab-practical.html#cb104-9"></a><span class="ss">        &gt; </span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/draft.out 2&gt; </span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/draft.err</span></span>
<span id="cb104-10"><a href="module-3---lab-practical.html#cb104-10"></a><span class="ss">&quot;&quot;&quot;</span>)</span></code></pre></div>
<p>Add additional reactions to enable growth on M9 minimal media.</p>
<div class="sourceCode" id="cb105"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb105-1"><a href="module-3---lab-practical.html#cb105-1"></a>Shell(<span class="ss">f&quot;&quot;&quot;</span></span>
<span id="cb105-2"><a href="module-3---lab-practical.html#cb105-2"></a><span class="ss">apptainer exec </span><span class="sc">{</span>gapseq_sif<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb105-3"><a href="module-3---lab-practical.html#cb105-3"></a><span class="ss">    gapseq fill --quick.gf </span><span class="op">\</span></span>
<span id="cb105-4"><a href="module-3---lab-practical.html#cb105-4"></a><span class="ss">        -m </span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/</span><span class="sc">{</span>gapseq_mag<span class="sc">}</span><span class="ss">-draft.RDS </span><span class="op">\</span></span>
<span id="cb105-5"><a href="module-3---lab-practical.html#cb105-5"></a><span class="ss">        -c </span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/</span><span class="sc">{</span>gapseq_mag<span class="sc">}</span><span class="ss">-rxnWeights.RDS </span><span class="op">\</span></span>
<span id="cb105-6"><a href="module-3---lab-practical.html#cb105-6"></a><span class="ss">        -g </span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/</span><span class="sc">{</span>gapseq_mag<span class="sc">}</span><span class="ss">-rxnXgenes.RDS </span><span class="op">\</span></span>
<span id="cb105-7"><a href="module-3---lab-practical.html#cb105-7"></a><span class="ss">        -n </span><span class="sc">{</span>LIB<span class="sc">}</span><span class="ss">/gapseq/m9.csv </span><span class="op">\</span></span>
<span id="cb105-8"><a href="module-3---lab-practical.html#cb105-8"></a><span class="ss">        --output.dir </span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss"> </span><span class="op">\</span></span>
<span id="cb105-9"><a href="module-3---lab-practical.html#cb105-9"></a><span class="ss">        &gt; </span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/fill.out 2&gt; </span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/fill.err</span></span>
<span id="cb105-10"><a href="module-3---lab-practical.html#cb105-10"></a><span class="ss">&quot;&quot;&quot;</span>)</span></code></pre></div>
<p>We can take a peek at the observed completion of each pathway.
The model will be examined further in the next lab.</p>
<div class="sourceCode" id="cb106"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb106-1"><a href="module-3---lab-practical.html#cb106-1"></a>Shell(<span class="ss">f&quot;&quot;&quot;</span></span>
<span id="cb106-2"><a href="module-3---lab-practical.html#cb106-2"></a><span class="ss">mag = 1</span></span>
<span id="cb106-3"><a href="module-3---lab-practical.html#cb106-3"></a><span class="ss">dfp = pd.read_csv(f&quot;./gapseq/mag</span><span class="sc">{</span>mag<span class="sc">}</span><span class="ss">/mag</span><span class="sc">{</span>mag<span class="sc">}</span><span class="ss">-all-Pathways.tbl&quot;, sep=&quot;</span><span class="ch">\t</span><span class="ss">&quot;, comment=&quot;#&quot;)</span></span>
<span id="cb106-4"><a href="module-3---lab-practical.html#cb106-4"></a><span class="ss">dfp.sort_values(&quot;Completeness&quot;, ascending=False, inplace=True)</span></span>
<span id="cb106-5"><a href="module-3---lab-practical.html#cb106-5"></a><span class="ss">dfp</span></span>
<span id="cb106-6"><a href="module-3---lab-practical.html#cb106-6"></a><span class="ss">&quot;&quot;&quot;</span>)</span></code></pre></div>
<p>The following code will generate a link to the KEGG pathway viewer with found reactions coloured.
Pick a few and have a look.</p>
<div class="sourceCode" id="cb107"><pre class="sourceCode numberSource python numberLines"><code class="sourceCode python"><span id="cb107-1"><a href="module-3---lab-practical.html#cb107-1"></a><span class="im">from</span> urllib.parse <span class="im">import</span> quote</span>
<span id="cb107-2"><a href="module-3---lab-practical.html#cb107-2"></a><span class="co"># https://www.genome.jp/kegg/webapp/color_url.html</span></span>
<span id="cb107-3"><a href="module-3---lab-practical.html#cb107-3"></a>pathway <span class="op">=</span> <span class="st">&quot;map00970&quot;</span></span>
<span id="cb107-4"><a href="module-3---lab-practical.html#cb107-4"></a>color <span class="op">=</span> <span class="st">&quot;#00cc96&quot;</span></span>
<span id="cb107-5"><a href="module-3---lab-practical.html#cb107-5"></a></span>
<span id="cb107-6"><a href="module-3---lab-practical.html#cb107-6"></a>reactions <span class="op">=</span> dfp[dfp[<span class="st">&quot;ID&quot;</span>] <span class="op">==</span> pathway][<span class="st">&quot;ReactionsFound&quot;</span>].tolist()</span>
<span id="cb107-7"><a href="module-3---lab-practical.html#cb107-7"></a>reactions <span class="op">=</span> [row.strip().split(<span class="st">&quot; &quot;</span>) <span class="cf">for</span> row <span class="kw">in</span> reactions]</span>
<span id="cb107-8"><a href="module-3---lab-practical.html#cb107-8"></a>reactions <span class="op">=</span> [<span class="ss">f&quot;</span><span class="sc">{</span>r<span class="sc">}</span><span class="ss"> </span><span class="sc">{</span>color<span class="sc">}</span><span class="ss">&quot;</span> <span class="cf">for</span> row <span class="kw">in</span> reactions <span class="cf">for</span> r <span class="kw">in</span> row]</span>
<span id="cb107-9"><a href="module-3---lab-practical.html#cb107-9"></a>q <span class="op">=</span> <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>.join(reactions)</span>
<span id="cb107-10"><a href="module-3---lab-practical.html#cb107-10"></a>url <span class="op">=</span> <span class="ss">f&quot;https://www.kegg.jp/kegg-bin/show_pathway?map=</span><span class="sc">{</span>pathway<span class="sc">}</span><span class="ss">&amp;multi_query=</span><span class="sc">{</span>quote(q)<span class="sc">}</span><span class="ss">&quot;</span></span>
<span id="cb107-11"><a href="module-3---lab-practical.html#cb107-11"></a>url</span></code></pre></div>
<p>This concludes the lab!</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="module-3-assigning-functions.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="module-4-visualization-and-finding-functional-significance.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
  "sharing": {
    "github": false,
    "facebook": true,
    "twitter": true,
    "linkedin": false,
    "weibo": false,
    "instapaper": false,
    "vk": false,
    "whatsapp": false,
    "all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
  },
  "fontsettings": {
    "theme": "white",
    "family": "sans",
    "size": 2
  },
  "edit": {
    "link": "https://github.com/cbw-dev/bookdown-template/blob/main/041-module-3-lab.Rmd",
    "text": "Edit"
  },
  "history": {
    "link": null,
    "text": null
  },
  "view": {
    "link": null,
    "text": null
  },
  "download": ["_main.pdf", "_main.epub"],
  "search": {
    "engine": "fuse",
    "options": null
  },
  "toc": {
    "collapse": "subsection"
  }
});
});
</script>

</body>

</html>
